{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "319f5eb9-3971-441f-82aa-116cea20430f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m24.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m24.3.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install -qU tiktoken pypdf openai langchain langchain-text-splitters langchain-openai langchain-core python-dotenv pydantic langchain_community"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4a95b19b-ff73-4f20-a874-e0842eb9131c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "import json\n",
    "from pprint import pprint\n",
    "from pathlib import Path\n",
    "import sys\n",
    "\n",
    "import tiktoken\n",
    "from pydantic import BaseModel, Field\n",
    "from openai import OpenAI\n",
    "import dotenv\n",
    "from langchain_community.document_loaders import PyPDFLoader\n",
    "from langchain_openai import ChatOpenAI\n",
    "from langchain_core.prompts import ChatPromptTemplate\n",
    "from langchain_text_splitters import RecursiveCharacterTextSplitter\n",
    "from langchain_core.documents.base import Document\n",
    "\n",
    "sys.path.append('..')\n",
    "from viime_extract.schema import ArticleMeta, ArticleKeyWords, Metabolite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4368b6b6-4be3-4076-9353-d7c46f1ba5e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "LLM_MODEL_NAME = 'gpt-4o-mini'\n",
    "PDF_DIR = Path('../data/PubMed LongCovid and Metabolomics Results')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5b87f2b5-ef10-4536-873f-e8c9362051cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "dotenv.load_dotenv()\n",
    "LLM = ChatOpenAI(model=LLM_MODEL_NAME, temperature=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "90e11497-c13a-4fdb-82d0-5b9018562b97",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pdf_pages(pdf_file: Path):\n",
    "    return list(PyPDFLoader(pdf_file).lazy_load())\n",
    "\n",
    "def extract_article_metadata(pages: list[Document], model=LLM):\n",
    "    # assumption: first page contains all of the metadata\n",
    "    first_page = pages[0].page_content\n",
    "\n",
    "    template = ChatPromptTemplate([\n",
    "        ('system', '''You are an expert in extracting structured information from medical journal articles.\n",
    "Present the extracted information in a clear, structured format. Be comprehensive and extract every single\n",
    "mentioned entity. You will be evaluated on the quality and completeness of the extracted information.'''),\n",
    "        ('user', 'Please extract the title, authors, journal title, publication year, journal volume, DOI ID, and pubmed ID from the following journal article.:\\n\\n{article_contents}'),\n",
    "    ])\n",
    "    prompt = template.invoke({'article_contents': first_page})\n",
    "    return model.with_structured_output(ArticleMeta).invoke(prompt)\n",
    "\n",
    "def extract_article_keywords(pages: list[Document], model=LLM, omit_bibliography=False):\n",
    "    template = ChatPromptTemplate([\n",
    "        ('system', '''You are an expert in extracting structured information from medical journal articles.\n",
    "Present the extracted information in a clear, structured format. Be comprehensive and extract every single\n",
    "mentioned entity. You will be evaluated on the quality and completeness of the extracted information.\n",
    "\n",
    "If you are not confident in the identifier for an entity, you can specify it as \"unknown\". It is better\n",
    "to include an entity with an \"unknown\" identified than to omit it entirely.'''),\n",
    "        ('user', 'Please extract the metabolites, proteins, genes, pathways, drugs, and diseases mentioned in the following journal article{prompt_suffix}:\\n\\n{article_contents}'),\n",
    "    ])\n",
    "\n",
    "    splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(\n",
    "        model_name=LLM_MODEL_NAME,\n",
    "        chunk_size=768,\n",
    "        chunk_overlap=64,\n",
    "    )\n",
    "\n",
    "    keywords = ArticleKeyWords()\n",
    "\n",
    "    prompt_suffix = ''\n",
    "    if omit_bibliography:\n",
    "        prompt_suffix = ', ignoring entities found in the references section'\n",
    "\n",
    "    chunks = splitter.split_documents(pages)\n",
    "    print('Number of chunks:', len(chunks))\n",
    "    for idx, text_chunk in enumerate(chunks):\n",
    "        print(f'Processing chunk #{idx}')\n",
    "        prompt = template.invoke({'article_contents': text_chunk.page_content, 'prompt_suffix': prompt_suffix})\n",
    "        response = model.with_structured_output(ArticleKeyWords).invoke(prompt)\n",
    "        keywords = keywords.merge(response)\n",
    "\n",
    "    return keywords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c28842a8-4836-4928-9060-88bebda36a39",
   "metadata": {},
   "outputs": [],
   "source": [
    "LC_81_pages = get_pdf_pages(PDF_DIR / 'LC_81.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "11aa8faa-7955-43f2-a2ca-79f4aebdb7ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "article_meta = extract_article_metadata(LC_81_pages, model=LLM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "74aa1995-ebea-414c-9730-6a08732750b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of chunks: 28\n",
      "Processing chunk #0\n",
      "Processing chunk #1\n",
      "Processing chunk #2\n",
      "Processing chunk #3\n",
      "Processing chunk #4\n",
      "Processing chunk #5\n",
      "Processing chunk #6\n",
      "Processing chunk #7\n",
      "Processing chunk #8\n",
      "Processing chunk #9\n",
      "Processing chunk #10\n",
      "Processing chunk #11\n",
      "Processing chunk #12\n",
      "Processing chunk #13\n",
      "Processing chunk #14\n",
      "Processing chunk #15\n",
      "Processing chunk #16\n",
      "Processing chunk #17\n",
      "Processing chunk #18\n",
      "Processing chunk #19\n",
      "Processing chunk #20\n",
      "Processing chunk #21\n",
      "Processing chunk #22\n",
      "Processing chunk #23\n",
      "Processing chunk #24\n",
      "Processing chunk #25\n",
      "Processing chunk #26\n",
      "Processing chunk #27\n",
      "CPU times: user 601 ms, sys: 77.8 ms, total: 679 ms\n",
      "Wall time: 1min 33s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "article_keywords = extract_article_keywords(LC_81_pages, model=LLM)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f84d55dd-6ec8-4358-a97b-19b3cc1e141e",
   "metadata": {},
   "source": [
    "# Ignoring references to the bibliography"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9e397451-121f-4346-8591-2aea18040859",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pydantic import BaseModel, Field\n",
    "from typing import Optional\n",
    "\n",
    "class Reference(BaseModel):\n",
    "    refnum: Optional[int] = Field(title=\"Reference Number\", description=\"The reference number\", default=None)\n",
    "    authors: Optional[list[str]] = Field(title=\"Authors\", description=\"List of authors in a reference\", default=None)\n",
    "    title: Optional[str] = Field(title=\"Paper title\", description=\"The reference paper title\", default=None)\n",
    "    year: Optional[int] = Field(title=\"Paper year\", description=\"The reference paper year of publication\", default=None)\n",
    "    journal: Optional[str] = Field(title=\"Journal\", description=\"The journal in which the reference was published\", default=None)\n",
    "\n",
    "    @property\n",
    "    def incomplete(self):\n",
    "        return any(getattr(self, prop) is None for prop in ['refnum', 'authors', 'title', 'year', 'journal'])\n",
    "\n",
    "class References(BaseModel):\n",
    "    references: Optional[list[Reference]] = Field(title=\"Bibliographic reference\", description=\"A bibliographic reference in an academic paper\", default=None)\n",
    "\n",
    "    @property\n",
    "    def complete_references(self):\n",
    "        return [r for r in self.references if not r.incomplete]\n",
    "\n",
    "def extract_references(page: Document, model=LLM):\n",
    "    template = ChatPromptTemplate([\n",
    "        ('system', '''You are an expert in extracting structured information from medical journal articles.\n",
    "Present the extracted information in a clear, structured format. Be comprehensive and extract every single reference. You will be evaluated on the quality and completeness of the extracted information. If you are unsure if an entry is a reference, then do not include it.'''),\n",
    "        ('user', 'Please list the references in this journal article:\\n\\n{article_contents}'),\n",
    "    ])\n",
    "    prompt = template.invoke({'article_contents': page})\n",
    "    return model.with_structured_output(References).invoke(prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "667bf564-b915-4b46-9c15-2450837a2e83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18\n"
     ]
    }
   ],
   "source": [
    "LC_69_pages = get_pdf_pages(PDF_DIR / 'LC_69.pdf')\n",
    "pages_up_to_bibliography = LC_69_pages[:13]\n",
    "print(len(LC_69_pages))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "78e2c4ec-1189-467b-a58a-a774c5733e84",
   "metadata": {},
   "outputs": [],
   "source": [
    "refs = extract_references(LC_69_pages[13])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "759f7438-b4fd-41a6-915d-7b965e319bef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Reference(refnum=1, authors=['B. Hu', 'S. Huang', 'L. Yin'], title='The cytokine storm and COVID-19', year=2021, journal='J. Med. Virol.'),\n",
       " Reference(refnum=2, authors=['C. Stasi', 'S. Fallani', 'F. Voller', 'C. Silvestri'], title='Treatment for COVID-19: an overview', year=2020, journal='Eur. J. Pharmacol.'),\n",
       " Reference(refnum=3, authors=['M.G. Netea', 'F. Balkwill', 'M. Chonchol', 'F. Cominelli', 'M.Y. Donath', 'E. J. Giamarellos-Bourboulis', 'D. Golenbock', 'M.S. Gresnigt', 'M.T. Heneka', 'H. M. Hoffman', 'R. Hotchkiss', 'L.A.B. Joosten', 'D.L. Kastner', 'M. Korte', 'E. Latz', 'P. Libby', 'T. Mandrup-Poulsen', 'A. Mantovani', 'K.H.G. Mills', 'K.L. Nowak', 'L. A. O ’ Neill', 'P. Pickkers', 'T. van der Poll', 'P.M. Ridker', 'J. Schalkwijk', 'D.A. Schwartz', 'B. Siegmund', 'C.J. Steer', 'H. Tilg', 'J.W.M. van der Meer', 'F.L. van de Veerdonk', 'C. A. Dinarello'], title='A guiding map for inflammation', year=2017, journal='Nat. Immunol.'),\n",
       " Reference(refnum=4, authors=['C.N. Serhan', 'P.A. Ward', 'D.W. Gilroy'], title='Fundamentals of Inflammation', year=2010, journal='Cambridge University Press'),\n",
       " Reference(refnum=5, authors=['J.Z. Haeggstrom'], title='Leukotriene biosynthetic enzymes as therapeutic targets', year=2018, journal='J. Clin. Invest.'),\n",
       " Reference(refnum=6, authors=['B. Samuelsson'], title='Leukotrienes: mediators of immediate hypersensitivity reactions and inflammation', year=1983, journal='Science'),\n",
       " Reference(refnum=7, authors=['C.N. Serhan'], title='Pro-resolving lipid mediators are leads for resolution physiology', year=2014, journal='Nature'),\n",
       " Reference(refnum=8, authors=['C.N. Serhan', 'S. Hong', 'K. Gronert', 'S.P. Colgan', 'P.R. Devchand', 'G. Mirick', 'R.-L. Moussignac'], title='Resolvins: a family of bioactive products of omega-3 fatty acid transformation circuits initiated by aspirin treatment that counter pro-inflammation signals', year=2002, journal='J. Exp. Med.'),\n",
       " Reference(refnum=9, authors=['C.N. Serhan', 'N. Chiang', 'T.E. Van Dyke'], title='Resolving inflammation: dual anti-inflammatory and pro-resolution lipid mediators', year=2008, journal='Nat. Rev. Immunol.'),\n",
       " Reference(refnum=10, authors=['L. Lindbom'], title='Regulation of vascular permeability by neutrophils in acute inflammation', year=2003, journal='Chem. Immunol. Allergy'),\n",
       " Reference(refnum=11, authors=['N. Chiang', 'G. Fredman', 'F. B ¨ackhed', 'S.F. Oh', 'T.W. Vickery', 'B.A. Schmidt', 'C. N. Serhan'], title='Infection regulates pro-resolving mediators that lower antibiotic requirements', year=2012, journal='Nature'),\n",
       " Reference(refnum=12, authors=['G.L. Bannenberg', 'N. Chiang', 'A. Ariel', 'M. Arita', 'E. Tjonahen', 'K.H. Gotlinger', 'S. Hong', 'C.N. Serhan'], title='Molecular circuits of resolution: formation and actions of resolvins and protectins', year=2005, journal='J. Immunol.'),\n",
       " Reference(refnum=13, authors=['B.D. Levy', 'C.B. Clish', 'B. Schmidt', 'K. Gronert', 'C.N. Serhan'], title='Lipid mediator class switching during acute inflammation: signals in resolution', year=2001, journal='Nat. Immunol.'),\n",
       " Reference(refnum=14, authors=['S. Hong', 'K. Gronert', 'P. Devchand', 'R.-L. Moussignac', 'C.N. Serhan'], title='Novel docosatrienes and 17S-resolvins generated from docosahexaenoic acid in murine brain, human blood and glial cells: autacoids in anti-inflammation', year=2003, journal='J. Biol. Chem.'),\n",
       " Reference(refnum=15, authors=['C.N. Serhan', 'R. Yang', 'K. Martinod', 'K. Kasuga', 'P.S. Pillai', 'T.F. Porter', 'S.F. Oh', 'M. Spite'], title='Maresins: novel macrophage mediators with potent anti-inflammatory and pro-resolving actions', year=2009, journal='J. Exp. Med.'),\n",
       " Reference(refnum=16, authors=['S. Arienti', 'N.D. Barth', 'D.A. Dorward', 'A.G. Rossi', 'I. Dransfield'], title='Regulation of apoptotic cell clearance during resolution of inflammation', year=2019, journal='Front. Pharmacol.'),\n",
       " Reference(refnum=17, authors=['J. Dalli', 'C.N. Serhan'], title='Specific lipid mediator signatures of human phagocytes: microparticles stimulate macrophage efferocytosis and pro-resolving mediators', year=2012, journal='Blood'),\n",
       " Reference(refnum=18, authors=['O. Werz', 'J. Gerstmeier', 'S. Libreros', 'X. De la Rosa', 'M. Werner', 'P.C. Norris', 'N. Chiang', 'C.N. Serhan'], title='Human macrophages differentially produce specific resolvin or leukotriene signals that depend on bacterial pathogenicity', year=2018, journal='Nat. Commun.'),\n",
       " Reference(refnum=19, authors=['P.M. Jordan', 'J. Gerstmeier', 'S. Pace', 'R. Bilancia', 'Z. Rao', 'F. Borner', 'L. Miek', 'O. Gutierrez-Gutierrez', 'V. Arakandy', 'A. Rossi', 'A. Ialenti', 'C. Gonzalez-Estevez', 'B. Loffler', 'L. Tuchscherr', 'C.N. Serhan', 'O. Werz'], title='Staphylococcus aureus-derived alpha-hemolysin evokes generation of specialized pro-resolving mediators promoting inflammation resolution', year=2020, journal='Cell Rep.'),\n",
       " Reference(refnum=20, authors=['T. Liu', 'A. Xiang', 'T. Peng', 'A.C. Doran', 'K.J. Tracey', 'B.J. Barnes', 'I. Tabas', 'M. Son', 'B. Diamond'], title='HMGB1-C1q complexes regulate macrophage function by switching between leukotriene and specialized proresolving mediator biosynthesis', year=2019, journal='Proc. Natl. Acad. Sci. U. S. A.'),\n",
       " Reference(refnum=21, authors=['C.N. Serhan', 'C.B. Clish', 'J. Brannon', 'S.P. Colgan', 'N. Chiang', 'K. Gronert'], title='Novel functional sets of lipid-derived mediators with antiinflammatory actions generated from omega-3 fatty acids via cyclooxygenase 2-nonsteroidal antiinflammatory drugs and transcellular processing', year=2000, journal='J. Exp. Med.'),\n",
       " Reference(refnum=22, authors=['N. Chiang', 'C.N. Serhan'], title='Cell-cell interaction in the transcellular biosynthesis of novel omega-3-derived lipid mediators', year=2006, journal='Methods Mol. Biol.'),\n",
       " Reference(refnum=23, authors=['A.S. Archambault', 'Y. Zaid', 'V. Rakotoarivelo', 'C. Turcotte', '´E. Dor ´e', 'I. Dubuc', 'C. Martin', 'O. Flamand', 'Y. Amar', 'A. Cheikh', 'H. Fares', 'A. El Hassani', 'Y. Tijani', 'A. C ˆot ´e', 'M. Laviolette', '´E. Boilard', 'L. Flamand', 'N. Flamand'], title='High levels of eicosanoids and docosanoids in the lungs of intubated COVID-19 patients', year=2021, journal='FASEB J.'),\n",
       " Reference(refnum=24, authors=['B. Schwarz', 'L. Sharma', 'L. Roberts', 'X. Peng', 'S. Bermejo', 'I. Leighton', 'A.C. Massana', 'S. Farhadian', 'A. Ko', 'C. DelaCruz', 'C.M. Bosio'], title='Severe SARS-CoV-2 infection in humans is defined by a shift in the serum lipidome resulting in dysregulation of eicosanoid immune mediators', year=2020, journal='medRxiv'),\n",
       " Reference(refnum=25, authors=['D. Koenis', 'I. Beegun', 'C. Jouvene', 'G.A. Aguirre', 'P.R. Souza', 'M. Gonzalez-Nunez', 'L. Ly', 'K. Pistorius', 'H.M. Kocher', 'W. Ricketts', 'G. Thomas', 'M. Perretti', 'G. Alusi', 'P. Pfeffer', 'J. Dalli'], title='Disrupted resolution mechanisms favor altered phagocyte responses in Covid-19', year=2021, journal='Circ. Res.'),\n",
       " Reference(refnum=26, authors=['J. Turnbull', 'R. Jha', 'C.A. Ortori', 'E. Lunt', 'P.J. Tighe', 'W.L. Irving', 'S.A. Gohir', 'D. H. Kim', 'A.M. Valdes', 'A.W. Tarr', 'D.A. Barrett', 'V. Chapman'], title='Serum levels of pro-inflammatory lipid mediators and specialised pro-resolving molecules are increased in SARS-CoV-2 patients and correlate with markers of the adaptive immune response', year=2022, journal='J. Infect. Dis.'),\n",
       " Reference(refnum=27, authors=['R. Yang', 'N. Chiang', 'S.F. Oh', 'C.N. Serhan'], title='Metabolomics-lipidomics of eicosanoids and docosanoids generated by phagocytes', year=2011, journal='Curr. Protoc. Immunol.'),\n",
       " Reference(refnum=28, authors=['C.N. Serhan', 'Y. Lu', 'S. Hong', 'R. Yang'], title='Mediator lipidomics: search algorithms for eicosanoids, resolvins and protectins', year=2007, journal='Meth. Enzymol.'),\n",
       " Reference(refnum=29, authors=['Y. Lu', 'S. Hong', 'R. Yang', 'J. Uddin', 'K.H. Gotlinger', 'N.A. Petasis', 'C.N. Serhan'], title='Identification of endogenous resolvin E1 and other lipid mediators derived from eicosapentaenoic acid via electrospray low energy tandem mass spectrometry: spectra and fragmentation mechanisms', year=2007, journal='Rapid Commun. Mass Spectrom.'),\n",
       " Reference(refnum=30, authors=['S. Hong', 'Y. Lu', 'R. Yang', 'K.H. Gotlinger', 'N.A. Petasis', 'C.N. Serhan'], title='Resolvin D1, protectin D1, and related docosahexaenoic acid-derived products: analysis via electrospray/low energy tandem mass spectrometry based on spectra and fragmentation mechanisms', year=2007, journal='J. Am. Soc. Mass Spectrom.'),\n",
       " Reference(refnum=31, authors=['B. Schwarz', 'L. Sharma', 'L. Roberts', 'X. Peng', 'S. Bermejo', 'I. Leighton', 'A. Casanovas-Massana', 'M. Minasyan', 'S. Farhadian', 'A.I. Ko', 'C.S. Dela Cruz', 'C.M. Bosio'], title='Cutting edge: severe SARS-CoV-2 infection in humans is defined by a shift in the serum lipidome, resulting in dysregulation of eicosanoid immune mediators', year=2021, journal='J. Immunol.'),\n",
       " Reference(refnum=32, authors=['E. Andreakos', 'M. Papadaki', 'C.N. Serhan'], title='Dexamethasone, pro-resolving lipid mediators and resolution of inflammation in COVID-19', year=2021, journal='Allergy.'),\n",
       " Reference(refnum=33, authors=['D. Panigrahy', 'M.M. Gilligan', 'S. Huang', 'A. Gartung', 'I. Cort ´es-Puch', 'P.J. Sime', 'R. P. Phipps', 'C.N. Serhan', 'B.D. Hammock'], title='Inflammation resolution: a dual-pronged approach to averting cytokine storms in COVID-19?', year=2020, journal='Cancer Metastasis Rev.'),\n",
       " Reference(refnum=34, authors=['A. Recchiuti', 'S. Patruno', 'D. Mattoscio', 'E. Isopi', 'A. Pomilio', 'A. Lamolinara', 'M. Iezzi', 'R. Pecce', 'M. Romano'], title='Resolvin D1 and D2 reduce SARS-CoV-2-induced inflammatory responses in cystic fibrosis macrophages', year=2021, journal='FASEB J.'),\n",
       " Reference(refnum=35, authors=['A.E. Barden', 'E. Mas', 'K.D. Croft', 'M. Phillips', 'T.A. Mori'], title='Specialized proresolving lipid mediators in humans with the metabolic syndrome after n-3 fatty acids and aspirin', year=2015, journal='Am. J. Clin. Nutr.'),\n",
       " Reference(refnum=36, authors=['H.A. Bazan', 'Y. Lu', 'B. Jun', 'Z. Fang', 'T.C. Woods', 'S. Hong'], title='Circulating inflammation-resolving lipid mediators RvD1 and DHA are decreased in patients with acutely symptomatic carotid disease', year=2017, journal='Prostaglandins Leukot. Essent. Fatty Acids.'),\n",
       " Reference(refnum=37, authors=['A.E. Barden', 'S. Shinde', 'V. Burke', 'I.B. Puddey', 'L.J. Beilin', 'A.B. Irish', 'G.F. Watts', 'T. A. Mori'], title='The effect of n-3 fatty acids and coenzyme Q10 supplementation on neutrophil leukotrienes, mediators of inflammation resolution and myeloperoxidase in chronic kidney disease', year=2018, journal='Prostaglandins Other Lipid Mediat.'),\n",
       " Reference(refnum=38, authors=['I. Hartling', 'A. Cremonesi', 'E. Osuna', 'P.H. Lou', 'E. Lucchinetti', 'M. Zaugg', 'M. Hersberger'], title='Quantitative profiling of inflammatory and pro-resolving lipid mediators in human adolescents and mouse plasma using UHPLC-MS/MS', year=2021, journal='Clin. Chem. Lab. Med.'),\n",
       " Reference(refnum=39, authors=['E. Mas', 'A. Barden', 'V. Burke', 'L.J. Beilin', 'G.F. Watts', 'R.C. Huang', 'I.B. Puddey', 'A. B. Irish', 'T.A. Mori'], title='A randomized controlled trial of the effects of n-3 fatty acids on resolvins in chronic kidney disease', year=2016, journal='Clin. Nutr. (Edinburgh, Scotland).'),\n",
       " Reference(refnum=40, authors=['E. Mas', 'K.D. Croft', 'P. Zahra', 'A. Barden', 'T.A. Mori'], title='Resolvins D.1, D2, and other mediators of self-limited resolution of inflammation in human blood following n-3 fatty acid supplementation', year=2012, journal='Clin. Chem.')]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refs.complete_references"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e532dc4a-d20c-4a15-a11a-3b8aa0e14bc6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====== 18\n",
      "27\n",
      "====== 17\n",
      "50\n",
      "====== 16\n",
      "47\n",
      "====== 15\n",
      "50\n",
      "====== 14\n",
      "40\n",
      "====== 13\n",
      "30\n",
      "====== 12\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[18], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m idx, page \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(LC_69_pages[::\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]):\n\u001b[1;32m      2\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m====== \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(LC_69_pages)\u001b[38;5;250m \u001b[39m\u001b[38;5;241m-\u001b[39m\u001b[38;5;250m \u001b[39midx\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m----> 3\u001b[0m     refs \u001b[38;5;241m=\u001b[39m \u001b[43mextract_references\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpage\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;28mlen\u001b[39m(refs\u001b[38;5;241m.\u001b[39mreferences))\n",
      "Cell \u001b[0;32mIn[17], line 20\u001b[0m, in \u001b[0;36mextract_references\u001b[0;34m(page, model)\u001b[0m\n\u001b[1;32m     14\u001b[0m     template \u001b[38;5;241m=\u001b[39m ChatPromptTemplate([\n\u001b[1;32m     15\u001b[0m         (\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msystem\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'''\u001b[39m\u001b[38;5;124mYou are an expert in extracting structured information from medical journal articles.\u001b[39m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;124mPresent the extracted information in a clear, structured format. Be comprehensive and extract every single reference. You will be evaluated on the quality and completeness of the extracted information. If you are unsure if an entry is a reference, then do not include it.\u001b[39m\u001b[38;5;124m'''\u001b[39m),\n\u001b[1;32m     17\u001b[0m         (\u001b[38;5;124m'\u001b[39m\u001b[38;5;124muser\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPlease list the references in this journal article:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m{article_contents}\u001b[39;00m\u001b[38;5;124m'\u001b[39m),\n\u001b[1;32m     18\u001b[0m     ])\n\u001b[1;32m     19\u001b[0m     prompt \u001b[38;5;241m=\u001b[39m template\u001b[38;5;241m.\u001b[39minvoke({\u001b[38;5;124m'\u001b[39m\u001b[38;5;124marticle_contents\u001b[39m\u001b[38;5;124m'\u001b[39m: page})\n\u001b[0;32m---> 20\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwith_structured_output\u001b[49m\u001b[43m(\u001b[49m\u001b[43mReferences\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minvoke\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprompt\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/langchain_core/runnables/base.py:3020\u001b[0m, in \u001b[0;36mRunnableSequence.invoke\u001b[0;34m(self, input, config, **kwargs)\u001b[0m\n\u001b[1;32m   3018\u001b[0m context\u001b[38;5;241m.\u001b[39mrun(_set_config_context, config)\n\u001b[1;32m   3019\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m i \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m-> 3020\u001b[0m     \u001b[38;5;28minput\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[43mcontext\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstep\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minvoke\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3021\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   3022\u001b[0m     \u001b[38;5;28minput\u001b[39m \u001b[38;5;241m=\u001b[39m context\u001b[38;5;241m.\u001b[39mrun(step\u001b[38;5;241m.\u001b[39minvoke, \u001b[38;5;28minput\u001b[39m, config)\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/langchain_core/runnables/base.py:5352\u001b[0m, in \u001b[0;36mRunnableBindingBase.invoke\u001b[0;34m(self, input, config, **kwargs)\u001b[0m\n\u001b[1;32m   5346\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21minvoke\u001b[39m(\n\u001b[1;32m   5347\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   5348\u001b[0m     \u001b[38;5;28minput\u001b[39m: Input,\n\u001b[1;32m   5349\u001b[0m     config: Optional[RunnableConfig] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m   5350\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Optional[Any],\n\u001b[1;32m   5351\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Output:\n\u001b[0;32m-> 5352\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbound\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minvoke\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   5353\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5354\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_merge_configs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconfig\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5355\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43m{\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   5356\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/langchain_core/language_models/chat_models.py:286\u001b[0m, in \u001b[0;36mBaseChatModel.invoke\u001b[0;34m(self, input, config, stop, **kwargs)\u001b[0m\n\u001b[1;32m    275\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21minvoke\u001b[39m(\n\u001b[1;32m    276\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    277\u001b[0m     \u001b[38;5;28minput\u001b[39m: LanguageModelInput,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    281\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[1;32m    282\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m BaseMessage:\n\u001b[1;32m    283\u001b[0m     config \u001b[38;5;241m=\u001b[39m ensure_config(config)\n\u001b[1;32m    284\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m cast(\n\u001b[1;32m    285\u001b[0m         ChatGeneration,\n\u001b[0;32m--> 286\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate_prompt\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    287\u001b[0m \u001b[43m            \u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_convert_input\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    288\u001b[0m \u001b[43m            \u001b[49m\u001b[43mstop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    289\u001b[0m \u001b[43m            \u001b[49m\u001b[43mcallbacks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcallbacks\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    290\u001b[0m \u001b[43m            \u001b[49m\u001b[43mtags\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtags\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    291\u001b[0m \u001b[43m            \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmetadata\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    292\u001b[0m \u001b[43m            \u001b[49m\u001b[43mrun_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mrun_name\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    293\u001b[0m \u001b[43m            \u001b[49m\u001b[43mrun_id\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpop\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mrun_id\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    294\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    295\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mgenerations[\u001b[38;5;241m0\u001b[39m][\u001b[38;5;241m0\u001b[39m],\n\u001b[1;32m    296\u001b[0m     )\u001b[38;5;241m.\u001b[39mmessage\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/langchain_core/language_models/chat_models.py:786\u001b[0m, in \u001b[0;36mBaseChatModel.generate_prompt\u001b[0;34m(self, prompts, stop, callbacks, **kwargs)\u001b[0m\n\u001b[1;32m    778\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mgenerate_prompt\u001b[39m(\n\u001b[1;32m    779\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    780\u001b[0m     prompts: \u001b[38;5;28mlist\u001b[39m[PromptValue],\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    783\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[1;32m    784\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m LLMResult:\n\u001b[1;32m    785\u001b[0m     prompt_messages \u001b[38;5;241m=\u001b[39m [p\u001b[38;5;241m.\u001b[39mto_messages() \u001b[38;5;28;01mfor\u001b[39;00m p \u001b[38;5;129;01min\u001b[39;00m prompts]\n\u001b[0;32m--> 786\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprompt_messages\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallbacks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallbacks\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/langchain_core/language_models/chat_models.py:643\u001b[0m, in \u001b[0;36mBaseChatModel.generate\u001b[0;34m(self, messages, stop, callbacks, tags, metadata, run_name, run_id, **kwargs)\u001b[0m\n\u001b[1;32m    641\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m run_managers:\n\u001b[1;32m    642\u001b[0m             run_managers[i]\u001b[38;5;241m.\u001b[39mon_llm_error(e, response\u001b[38;5;241m=\u001b[39mLLMResult(generations\u001b[38;5;241m=\u001b[39m[]))\n\u001b[0;32m--> 643\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    644\u001b[0m flattened_outputs \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m    645\u001b[0m     LLMResult(generations\u001b[38;5;241m=\u001b[39m[res\u001b[38;5;241m.\u001b[39mgenerations], llm_output\u001b[38;5;241m=\u001b[39mres\u001b[38;5;241m.\u001b[39mllm_output)  \u001b[38;5;66;03m# type: ignore[list-item]\u001b[39;00m\n\u001b[1;32m    646\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m results\n\u001b[1;32m    647\u001b[0m ]\n\u001b[1;32m    648\u001b[0m llm_output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_combine_llm_outputs([res\u001b[38;5;241m.\u001b[39mllm_output \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m results])\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/langchain_core/language_models/chat_models.py:633\u001b[0m, in \u001b[0;36mBaseChatModel.generate\u001b[0;34m(self, messages, stop, callbacks, tags, metadata, run_name, run_id, **kwargs)\u001b[0m\n\u001b[1;32m    630\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(messages):\n\u001b[1;32m    631\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    632\u001b[0m         results\u001b[38;5;241m.\u001b[39mappend(\n\u001b[0;32m--> 633\u001b[0m             \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_generate_with_cache\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    634\u001b[0m \u001b[43m                \u001b[49m\u001b[43mm\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    635\u001b[0m \u001b[43m                \u001b[49m\u001b[43mstop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    636\u001b[0m \u001b[43m                \u001b[49m\u001b[43mrun_manager\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrun_managers\u001b[49m\u001b[43m[\u001b[49m\u001b[43mi\u001b[49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mrun_managers\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    637\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    638\u001b[0m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    639\u001b[0m         )\n\u001b[1;32m    640\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    641\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m run_managers:\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/langchain_core/language_models/chat_models.py:851\u001b[0m, in \u001b[0;36mBaseChatModel._generate_with_cache\u001b[0;34m(self, messages, stop, run_manager, **kwargs)\u001b[0m\n\u001b[1;32m    849\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    850\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m inspect\u001b[38;5;241m.\u001b[39msignature(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_generate)\u001b[38;5;241m.\u001b[39mparameters\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrun_manager\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[0;32m--> 851\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_generate\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    852\u001b[0m \u001b[43m            \u001b[49m\u001b[43mmessages\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrun_manager\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrun_manager\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\n\u001b[1;32m    853\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    854\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    855\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_generate(messages, stop\u001b[38;5;241m=\u001b[39mstop, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/langchain_openai/chat_models/base.py:773\u001b[0m, in \u001b[0;36mBaseChatOpenAI._generate\u001b[0;34m(self, messages, stop, run_manager, **kwargs)\u001b[0m\n\u001b[1;32m    771\u001b[0m payload\u001b[38;5;241m.\u001b[39mpop(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstream\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    772\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 773\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mroot_client\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbeta\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mchat\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcompletions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparse\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mpayload\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    774\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m openai\u001b[38;5;241m.\u001b[39mBadRequestError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    775\u001b[0m     _handle_openai_bad_request(e)\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/openai/resources/beta/chat/completions.py:160\u001b[0m, in \u001b[0;36mCompletions.parse\u001b[0;34m(self, messages, model, audio, response_format, frequency_penalty, function_call, functions, logit_bias, logprobs, max_completion_tokens, max_tokens, metadata, modalities, n, parallel_tool_calls, prediction, presence_penalty, reasoning_effort, seed, service_tier, stop, store, stream_options, temperature, tool_choice, tools, top_logprobs, top_p, user, extra_headers, extra_query, extra_body, timeout)\u001b[0m\n\u001b[1;32m    153\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mparser\u001b[39m(raw_completion: ChatCompletion) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ParsedChatCompletion[ResponseFormatT]:\n\u001b[1;32m    154\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _parse_chat_completion(\n\u001b[1;32m    155\u001b[0m         response_format\u001b[38;5;241m=\u001b[39mresponse_format,\n\u001b[1;32m    156\u001b[0m         chat_completion\u001b[38;5;241m=\u001b[39mraw_completion,\n\u001b[1;32m    157\u001b[0m         input_tools\u001b[38;5;241m=\u001b[39mtools,\n\u001b[1;32m    158\u001b[0m     )\n\u001b[0;32m--> 160\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_post\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    161\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/chat/completions\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    162\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmaybe_transform\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    163\u001b[0m \u001b[43m        \u001b[49m\u001b[43m{\u001b[49m\n\u001b[1;32m    164\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmessages\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmessages\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    165\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmodel\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    166\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43maudio\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43maudio\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    167\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfrequency_penalty\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfrequency_penalty\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    168\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfunction_call\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfunction_call\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    169\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfunctions\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfunctions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    170\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mlogit_bias\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mlogit_bias\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    171\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mlogprobs\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mlogprobs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    172\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmax_completion_tokens\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmax_completion_tokens\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    173\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmax_tokens\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmax_tokens\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    174\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmetadata\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    175\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmodalities\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmodalities\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    176\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mn\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    177\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mparallel_tool_calls\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mparallel_tool_calls\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    178\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mprediction\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mprediction\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    179\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpresence_penalty\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mpresence_penalty\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    180\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mreasoning_effort\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mreasoning_effort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    181\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mresponse_format\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43m_type_to_response_format\u001b[49m\u001b[43m(\u001b[49m\u001b[43mresponse_format\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    182\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mseed\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mseed\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    183\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mservice_tier\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mservice_tier\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    184\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstop\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    185\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstore\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mstore\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    186\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstream\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    187\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstream_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mstream_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    188\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtemperature\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mtemperature\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    189\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtool_choice\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mtool_choice\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    190\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtools\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mtools\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    191\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtop_logprobs\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mtop_logprobs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    192\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtop_p\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mtop_p\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    193\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43muser\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43muser\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    194\u001b[0m \u001b[43m        \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    195\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcompletion_create_params\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mCompletionCreateParams\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    196\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    197\u001b[0m \u001b[43m    \u001b[49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmake_request_options\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    198\u001b[0m \u001b[43m        \u001b[49m\u001b[43mextra_headers\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mextra_headers\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    199\u001b[0m \u001b[43m        \u001b[49m\u001b[43mextra_query\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mextra_query\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    200\u001b[0m \u001b[43m        \u001b[49m\u001b[43mextra_body\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mextra_body\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    201\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    202\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpost_parser\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparser\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    203\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    204\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# we turn the `ChatCompletion` instance into a `ParsedChatCompletion`\u001b[39;49;00m\n\u001b[1;32m    205\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# in the `parser` function above\u001b[39;49;00m\n\u001b[1;32m    206\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcast_to\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcast\u001b[49m\u001b[43m(\u001b[49m\u001b[43mType\u001b[49m\u001b[43m[\u001b[49m\u001b[43mParsedChatCompletion\u001b[49m\u001b[43m[\u001b[49m\u001b[43mResponseFormatT\u001b[49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mChatCompletion\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    207\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstream\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    208\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/openai/_base_client.py:1283\u001b[0m, in \u001b[0;36mSyncAPIClient.post\u001b[0;34m(self, path, cast_to, body, options, files, stream, stream_cls)\u001b[0m\n\u001b[1;32m   1269\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mpost\u001b[39m(\n\u001b[1;32m   1270\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   1271\u001b[0m     path: \u001b[38;5;28mstr\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1278\u001b[0m     stream_cls: \u001b[38;5;28mtype\u001b[39m[_StreamT] \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m   1279\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ResponseT \u001b[38;5;241m|\u001b[39m _StreamT:\n\u001b[1;32m   1280\u001b[0m     opts \u001b[38;5;241m=\u001b[39m FinalRequestOptions\u001b[38;5;241m.\u001b[39mconstruct(\n\u001b[1;32m   1281\u001b[0m         method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpost\u001b[39m\u001b[38;5;124m\"\u001b[39m, url\u001b[38;5;241m=\u001b[39mpath, json_data\u001b[38;5;241m=\u001b[39mbody, files\u001b[38;5;241m=\u001b[39mto_httpx_files(files), \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39moptions\n\u001b[1;32m   1282\u001b[0m     )\n\u001b[0;32m-> 1283\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m cast(ResponseT, \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcast_to\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mopts\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstream\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstream\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstream_cls\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstream_cls\u001b[49m\u001b[43m)\u001b[49m)\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/openai/_base_client.py:960\u001b[0m, in \u001b[0;36mSyncAPIClient.request\u001b[0;34m(self, cast_to, options, remaining_retries, stream, stream_cls)\u001b[0m\n\u001b[1;32m    957\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    958\u001b[0m     retries_taken \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[0;32m--> 960\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    961\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcast_to\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcast_to\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    962\u001b[0m \u001b[43m    \u001b[49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    963\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstream\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstream\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    964\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstream_cls\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstream_cls\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    965\u001b[0m \u001b[43m    \u001b[49m\u001b[43mretries_taken\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretries_taken\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    966\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/openai/_base_client.py:996\u001b[0m, in \u001b[0;36mSyncAPIClient._request\u001b[0;34m(self, cast_to, options, retries_taken, stream, stream_cls)\u001b[0m\n\u001b[1;32m    993\u001b[0m log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSending HTTP Request: \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, request\u001b[38;5;241m.\u001b[39mmethod, request\u001b[38;5;241m.\u001b[39murl)\n\u001b[1;32m    995\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 996\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_client\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    997\u001b[0m \u001b[43m        \u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    998\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstream\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstream\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_should_stream_response_body\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    999\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1000\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1001\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m httpx\u001b[38;5;241m.\u001b[39mTimeoutException \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m   1002\u001b[0m     log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mEncountered httpx.TimeoutException\u001b[39m\u001b[38;5;124m\"\u001b[39m, exc_info\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/httpx/_client.py:914\u001b[0m, in \u001b[0;36mClient.send\u001b[0;34m(self, request, stream, auth, follow_redirects)\u001b[0m\n\u001b[1;32m    910\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_set_timeout(request)\n\u001b[1;32m    912\u001b[0m auth \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_build_request_auth(request, auth)\n\u001b[0;32m--> 914\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_handling_auth\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    915\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    916\u001b[0m \u001b[43m    \u001b[49m\u001b[43mauth\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mauth\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    917\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfollow_redirects\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfollow_redirects\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    918\u001b[0m \u001b[43m    \u001b[49m\u001b[43mhistory\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    919\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    920\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    921\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m stream:\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/httpx/_client.py:942\u001b[0m, in \u001b[0;36mClient._send_handling_auth\u001b[0;34m(self, request, auth, follow_redirects, history)\u001b[0m\n\u001b[1;32m    939\u001b[0m request \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mnext\u001b[39m(auth_flow)\n\u001b[1;32m    941\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 942\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_handling_redirects\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    943\u001b[0m \u001b[43m        \u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    944\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfollow_redirects\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfollow_redirects\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    945\u001b[0m \u001b[43m        \u001b[49m\u001b[43mhistory\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhistory\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    946\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    947\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    948\u001b[0m         \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/httpx/_client.py:979\u001b[0m, in \u001b[0;36mClient._send_handling_redirects\u001b[0;34m(self, request, follow_redirects, history)\u001b[0m\n\u001b[1;32m    976\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m hook \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_event_hooks[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrequest\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[1;32m    977\u001b[0m     hook(request)\n\u001b[0;32m--> 979\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_single_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    980\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    981\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m hook \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_event_hooks[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresponse\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/httpx/_client.py:1014\u001b[0m, in \u001b[0;36mClient._send_single_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m   1009\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\n\u001b[1;32m   1010\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAttempted to send an async request with a sync Client instance.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1011\u001b[0m     )\n\u001b[1;32m   1013\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m request_context(request\u001b[38;5;241m=\u001b[39mrequest):\n\u001b[0;32m-> 1014\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mtransport\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1016\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(response\u001b[38;5;241m.\u001b[39mstream, SyncByteStream)\n\u001b[1;32m   1018\u001b[0m response\u001b[38;5;241m.\u001b[39mrequest \u001b[38;5;241m=\u001b[39m request\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/httpx/_transports/default.py:250\u001b[0m, in \u001b[0;36mHTTPTransport.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    237\u001b[0m req \u001b[38;5;241m=\u001b[39m httpcore\u001b[38;5;241m.\u001b[39mRequest(\n\u001b[1;32m    238\u001b[0m     method\u001b[38;5;241m=\u001b[39mrequest\u001b[38;5;241m.\u001b[39mmethod,\n\u001b[1;32m    239\u001b[0m     url\u001b[38;5;241m=\u001b[39mhttpcore\u001b[38;5;241m.\u001b[39mURL(\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    247\u001b[0m     extensions\u001b[38;5;241m=\u001b[39mrequest\u001b[38;5;241m.\u001b[39mextensions,\n\u001b[1;32m    248\u001b[0m )\n\u001b[1;32m    249\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m map_httpcore_exceptions():\n\u001b[0;32m--> 250\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_pool\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreq\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    252\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(resp\u001b[38;5;241m.\u001b[39mstream, typing\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m    254\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m Response(\n\u001b[1;32m    255\u001b[0m     status_code\u001b[38;5;241m=\u001b[39mresp\u001b[38;5;241m.\u001b[39mstatus,\n\u001b[1;32m    256\u001b[0m     headers\u001b[38;5;241m=\u001b[39mresp\u001b[38;5;241m.\u001b[39mheaders,\n\u001b[1;32m    257\u001b[0m     stream\u001b[38;5;241m=\u001b[39mResponseStream(resp\u001b[38;5;241m.\u001b[39mstream),\n\u001b[1;32m    258\u001b[0m     extensions\u001b[38;5;241m=\u001b[39mresp\u001b[38;5;241m.\u001b[39mextensions,\n\u001b[1;32m    259\u001b[0m )\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/httpcore/_sync/connection_pool.py:256\u001b[0m, in \u001b[0;36mConnectionPool.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    253\u001b[0m         closing \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_assign_requests_to_connections()\n\u001b[1;32m    255\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_close_connections(closing)\n\u001b[0;32m--> 256\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exc \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    258\u001b[0m \u001b[38;5;66;03m# Return the response. Note that in this case we still have to manage\u001b[39;00m\n\u001b[1;32m    259\u001b[0m \u001b[38;5;66;03m# the point at which the response is closed.\u001b[39;00m\n\u001b[1;32m    260\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(response\u001b[38;5;241m.\u001b[39mstream, typing\u001b[38;5;241m.\u001b[39mIterable)\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/httpcore/_sync/connection_pool.py:236\u001b[0m, in \u001b[0;36mConnectionPool.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    232\u001b[0m connection \u001b[38;5;241m=\u001b[39m pool_request\u001b[38;5;241m.\u001b[39mwait_for_connection(timeout\u001b[38;5;241m=\u001b[39mtimeout)\n\u001b[1;32m    234\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    235\u001b[0m     \u001b[38;5;66;03m# Send the request on the assigned connection.\u001b[39;00m\n\u001b[0;32m--> 236\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mconnection\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    237\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpool_request\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\n\u001b[1;32m    238\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    239\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m ConnectionNotAvailable:\n\u001b[1;32m    240\u001b[0m     \u001b[38;5;66;03m# In some cases a connection may initially be available to\u001b[39;00m\n\u001b[1;32m    241\u001b[0m     \u001b[38;5;66;03m# handle a request, but then become unavailable.\u001b[39;00m\n\u001b[1;32m    242\u001b[0m     \u001b[38;5;66;03m#\u001b[39;00m\n\u001b[1;32m    243\u001b[0m     \u001b[38;5;66;03m# In this case we clear the connection and try again.\u001b[39;00m\n\u001b[1;32m    244\u001b[0m     pool_request\u001b[38;5;241m.\u001b[39mclear_connection()\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/httpcore/_sync/connection.py:103\u001b[0m, in \u001b[0;36mHTTPConnection.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    100\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_connect_failed \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    101\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exc\n\u001b[0;32m--> 103\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_connection\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/httpcore/_sync/http11.py:136\u001b[0m, in \u001b[0;36mHTTP11Connection.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    134\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m Trace(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresponse_closed\u001b[39m\u001b[38;5;124m\"\u001b[39m, logger, request) \u001b[38;5;28;01mas\u001b[39;00m trace:\n\u001b[1;32m    135\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_response_closed()\n\u001b[0;32m--> 136\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m exc\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/httpcore/_sync/http11.py:106\u001b[0m, in \u001b[0;36mHTTP11Connection.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m     95\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n\u001b[1;32m     97\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m Trace(\n\u001b[1;32m     98\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreceive_response_headers\u001b[39m\u001b[38;5;124m\"\u001b[39m, logger, request, kwargs\n\u001b[1;32m     99\u001b[0m ) \u001b[38;5;28;01mas\u001b[39;00m trace:\n\u001b[1;32m    100\u001b[0m     (\n\u001b[1;32m    101\u001b[0m         http_version,\n\u001b[1;32m    102\u001b[0m         status,\n\u001b[1;32m    103\u001b[0m         reason_phrase,\n\u001b[1;32m    104\u001b[0m         headers,\n\u001b[1;32m    105\u001b[0m         trailing_data,\n\u001b[0;32m--> 106\u001b[0m     ) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_receive_response_headers\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    107\u001b[0m     trace\u001b[38;5;241m.\u001b[39mreturn_value \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    108\u001b[0m         http_version,\n\u001b[1;32m    109\u001b[0m         status,\n\u001b[1;32m    110\u001b[0m         reason_phrase,\n\u001b[1;32m    111\u001b[0m         headers,\n\u001b[1;32m    112\u001b[0m     )\n\u001b[1;32m    114\u001b[0m network_stream \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_network_stream\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/httpcore/_sync/http11.py:177\u001b[0m, in \u001b[0;36mHTTP11Connection._receive_response_headers\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    174\u001b[0m timeout \u001b[38;5;241m=\u001b[39m timeouts\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mread\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m    176\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 177\u001b[0m     event \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_receive_event\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    178\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(event, h11\u001b[38;5;241m.\u001b[39mResponse):\n\u001b[1;32m    179\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/httpcore/_sync/http11.py:217\u001b[0m, in \u001b[0;36mHTTP11Connection._receive_event\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    214\u001b[0m     event \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_h11_state\u001b[38;5;241m.\u001b[39mnext_event()\n\u001b[1;32m    216\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m event \u001b[38;5;129;01mis\u001b[39;00m h11\u001b[38;5;241m.\u001b[39mNEED_DATA:\n\u001b[0;32m--> 217\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_network_stream\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    218\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mREAD_NUM_BYTES\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\n\u001b[1;32m    219\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    221\u001b[0m     \u001b[38;5;66;03m# If we feed this case through h11 we'll raise an exception like:\u001b[39;00m\n\u001b[1;32m    222\u001b[0m     \u001b[38;5;66;03m#\u001b[39;00m\n\u001b[1;32m    223\u001b[0m     \u001b[38;5;66;03m#     httpcore.RemoteProtocolError: can't handle event type\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    227\u001b[0m     \u001b[38;5;66;03m# perspective. Instead we handle this case distinctly and treat\u001b[39;00m\n\u001b[1;32m    228\u001b[0m     \u001b[38;5;66;03m# it as a ConnectError.\u001b[39;00m\n\u001b[1;32m    229\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m data \u001b[38;5;241m==\u001b[39m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_h11_state\u001b[38;5;241m.\u001b[39mtheir_state \u001b[38;5;241m==\u001b[39m h11\u001b[38;5;241m.\u001b[39mSEND_RESPONSE:\n",
      "File \u001b[0;32m~/code/viime/viime-extract/venv/lib/python3.13/site-packages/httpcore/_backends/sync.py:128\u001b[0m, in \u001b[0;36mSyncStream.read\u001b[0;34m(self, max_bytes, timeout)\u001b[0m\n\u001b[1;32m    126\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m map_exceptions(exc_map):\n\u001b[1;32m    127\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sock\u001b[38;5;241m.\u001b[39msettimeout(timeout)\n\u001b[0;32m--> 128\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmax_bytes\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/homebrew/Cellar/python@3.13/3.13.0_1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/ssl.py:1285\u001b[0m, in \u001b[0;36mSSLSocket.recv\u001b[0;34m(self, buflen, flags)\u001b[0m\n\u001b[1;32m   1281\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m flags \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m   1282\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1283\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnon-zero flags not allowed in calls to recv() on \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m\n\u001b[1;32m   1284\u001b[0m             \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m)\n\u001b[0;32m-> 1285\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbuflen\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1286\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1287\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mrecv(buflen, flags)\n",
      "File \u001b[0;32m/opt/homebrew/Cellar/python@3.13/3.13.0_1/Frameworks/Python.framework/Versions/3.13/lib/python3.13/ssl.py:1140\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1138\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sslobj\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;28mlen\u001b[39m, buffer)\n\u001b[1;32m   1139\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1140\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sslobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1141\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m SSLError \u001b[38;5;28;01mas\u001b[39;00m x:\n\u001b[1;32m   1142\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m x\u001b[38;5;241m.\u001b[39margs[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m==\u001b[39m SSL_ERROR_EOF \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msuppress_ragged_eofs:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "for idx, page in enumerate(LC_69_pages[::-1]):\n",
    "    print(f'====== {len(LC_69_pages) - idx}')\n",
    "    refs = extract_references(page)\n",
    "    print(len(refs.references))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b2129d67-ac0d-495f-adf5-e5c8553951e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "References(references=[Reference(refnum=8, authors=[], title='Acid-methanol trapping studies', year=2022), Reference(refnum=14, authors=[], title='Studies with leukocytes, isolated enzymes, and their stereoselective total organic synthesis', year=2022), Reference(refnum=15, authors=[], title='Evidence for the cellular production of unstable epoxide intermediates', year=2022), Reference(refnum=27, authors=[], title='Targeted LC-MS-MS-based profiling', year=2022), Reference(refnum=28, authors=[], title='Targeted LC-MS-MS-based metabololipidomic profiling', year=2022), Reference(refnum=38, authors=[], title='SPM biosynthesis in vivo in humans', year=2022), Reference(refnum=39, authors=[], title='SPM biosynthesis in vivo in humans', year=2022), Reference(refnum=40, authors=[], title='SPM biosynthesis in vivo in humans', year=2022), Reference(refnum=50, authors=[], title='Omega-3 supplementation and SPM production', year=2022), Reference(refnum=63, authors=[], title='Clinical trial with proresolving mimetic LX/Rv stable analog', year=2022), Reference(refnum=75, authors=[], title='Functional roles for SPMs in peripheral blood circulation', year=2022), Reference(refnum=80, authors=[], title='SPMs activate endogenous resolution programs', year=2022), Reference(refnum=142, authors=[], title='Cell type specific enzymatic conversion to Resolvin D3 and Resolvin D4', year=2022), Reference(refnum=174, authors=[], title='Review on the biosynthesis of Resolvins, Protectins and Maresins', year=2022), Reference(refnum=176, authors=['Nicos Petasis'], title='Total organic synthesis of 13 S,14 S-epoxy-maresin', year=2022), Reference(refnum=177, authors=[], title='Conversion to MaR1 by M2 macrophages', year=2022), Reference(refnum=178, authors=['Trond Hansen', 'Jørn Tungen', 'Marius Aursnes'], title='Preparation of 16,17-epoxy-protectin', year=2022), Reference(refnum=179, authors=[], title='MCTRs stimulate tissue regeneration', year=2022), Reference(refnum=180, authors=[], title='Complete stereochemistry and bioactivities of SPMs', year=2022), Reference(refnum=181, authors=['Bernd Spur', 'Nicos Petasis'], title='Studies on SPMs', year=2022), Reference(refnum=182, authors=[], title='Studies on SPMs', year=2022), Reference(refnum=183, authors=[], title='Enzymes responsible for conversion to potent cys-SPMs', year=2022), Reference(refnum=184, authors=[], title='Review on stereoselective biosynthesis of resolvins', year=2022), Reference(refnum=190, authors=[], title='Clinical studies with precision nutrition', year=2022), Reference(refnum=191, authors=[], title='Local delivery of pro-resolving molecules', year=2022), Reference(refnum=192, authors=[], title='Engineered materials for activating endogenous resolution programs', year=2022), Reference(refnum=193, authors=[], title='Engineered materials for activating endogenous resolution programs', year=2022), Reference(refnum=188, authors=[], title='Protective function of SPMs', year=2022), Reference(refnum=189, authors=[], title='Importance of precision nutrition', year=2022), Reference(refnum=1420, authors=[], title='Publications on resolvins in PubMed', year=2022)])"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "590b66d4-019b-4a72-ad16-b33f4159cfae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seminars in Immunology 59 (2022) 101597\n",
      "13\n",
      "in human public health and the potential for SPMs in resolution \n",
      "physiology-pharmacology. \n",
      "Author contributions \n",
      "C.N.S., S.L. and R.N. composed and contributed to the preparation of \n",
      "this manuscript, tables, and figures. \n",
      "Funding \n",
      "C.N.S is supported by the National Institutes of Health (grant no. \n",
      "R35GM095467) and S.L. is supported by NIH grant no. K99HL153673. \n",
      "Acknowledgments \n",
      "We thank M. H. Small for expert assistance in manuscript prepara -\n",
      "tion. We thank the many investigators that have published exciting re -\n",
      "sults on the potent functions and pharmacology of the SPMs and their \n",
      "total organic syntheses reviewed herein. \n",
      "References \n",
      "[1] B. Hu, S. Huang, L. Yin, The cytokine storm and COVID-19, J. Med. Virol. 93 (1) \n",
      "(2021) 250 – 256 . \n",
      "[2] C. Stasi, S. Fallani, F. Voller, C. Silvestri, Treatment for COVID-19: an overview, \n",
      "Eur. J. Pharmacol. 889 (2020), 173644 . \n",
      "[3] M.G. Netea, F. Balkwill, M. Chonchol, F. Cominelli, M.Y. Donath, E. \n",
      "J. Giamarellos-Bourboulis, D. Golenbock, M.S. Gresnigt, M.T. Heneka, H. \n",
      "M. Hoffman, R. Hotchkiss, L.A.B. Joosten, D.L. Kastner, M. Korte, E. Latz, \n",
      "P. Libby, T. Mandrup-Poulsen, A. Mantovani, K.H.G. Mills, K.L. Nowak, L. \n",
      "A. O ’ Neill, P. Pickkers, T. van der Poll, P.M. Ridker, J. Schalkwijk, D.A. Schwartz, \n",
      "B. Siegmund, C.J. Steer, H. Tilg, J.W.M. van der Meer, F.L. van de Veerdonk, C. \n",
      "A. Dinarello, A guiding map for inflammation, Nat. Immunol. 18 (8) (2017) \n",
      "826 – 831 . \n",
      "[4] C.N. Serhan, P.A. Ward, D.W. Gilroy, Fundamentals of Inflammation, Cambridge \n",
      "University Press, New York, 2010 . \n",
      "[5] J.Z. Haeggstrom, Leukotriene biosynthetic enzymes as therapeutic targets, \n",
      "J. Clin. Invest. 128 (7) (2018) 2680 – 2690 . \n",
      "[6] B. Samuelsson, Leukotrienes: mediators of immediate hypersensitivity reactions \n",
      "and inflammation, Science 220 (1983) 568 – 575 . \n",
      "[7] C.N. Serhan, Pro-resolving lipid mediators are leads for resolution physiology, \n",
      "Nature 510 (7503) (2014) 92 – 101 . \n",
      "[8] C.N. Serhan, S. Hong, K. Gronert, S.P. Colgan, P.R. Devchand, G. Mirick, R.- \n",
      "L. Moussignac, Resolvins: a family of bioactive products of omega-3 fatty acid \n",
      "transformation circuits initiated by aspirin treatment that counter pro- \n",
      "inflammation signals, J. Exp. Med. 196 (2002) 1025 – 1037 . \n",
      "[9] C.N. Serhan, N. Chiang, T.E. Van Dyke, Resolving inflammation: dual anti- \n",
      "inflammatory and pro-resolution lipid mediators, Nat. Rev. Immunol. 8 (2008) \n",
      "349 – 361 . \n",
      "[10] L. Lindbom, Regulation of vascular permeability by neutrophils in acute \n",
      "inflammation, Chem. Immunol. Allergy 83 (2003) 146 – 166 . \n",
      "[11] N. Chiang, G. Fredman, F. B ¨ackhed, S.F. Oh, T.W. Vickery, B.A. Schmidt, C. \n",
      "N. Serhan, Infection regulates pro-resolving mediators that lower antibiotic \n",
      "requirements, Nature 484 (2012) 524 – 528 . \n",
      "[12] G.L. Bannenberg, N. Chiang, A. Ariel, M. Arita, E. Tjonahen, K.H. Gotlinger, \n",
      "S. Hong, C.N. Serhan, Molecular circuits of resolution: formation and actions of \n",
      "resolvins and protectins, J. Immunol. 174 (2005) 4345 – 4355 . \n",
      "[13] B.D. Levy, C.B. Clish, B. Schmidt, K. Gronert, C.N. Serhan, Lipid mediator class \n",
      "switching during acute inflammation: signals in resolution, Nat. Immunol. 2 \n",
      "(2001) 612 – 619 . \n",
      "[14] S. Hong, K. Gronert, P. Devchand, R.-L. Moussignac, C.N. Serhan, Novel \n",
      "docosatrienes and 17S-resolvins generated from docosahexaenoic acid in murine \n",
      "brain, human blood and glial cells: autacoids in anti-inflammation, J. Biol. Chem. \n",
      "278 (2003) 14677 – 14687 . \n",
      "[15] C.N. Serhan, R. Yang, K. Martinod, K. Kasuga, P.S. Pillai, T.F. Porter, S.F. Oh, \n",
      "M. Spite, Maresins: novel macrophage mediators with potent anti-inflammatory \n",
      "and pro-resolving actions, J. Exp. Med. 206 (2009) 15 – 23 . \n",
      "[16] S. Arienti, N.D. Barth, D.A. Dorward, A.G. Rossi, I. Dransfield, Regulation of \n",
      "apoptotic cell clearance during resolution of inflammation, Front. Pharmacol. 10 \n",
      "(2019) 891 . \n",
      "[17] J. Dalli, C.N. Serhan, Specific lipid mediator signatures of human phagocytes: \n",
      "microparticles stimulate macrophage efferocytosis and pro-resolving mediators, \n",
      "Blood 120 (2012) e60 – 72 . \n",
      "[18] O. Werz, J. Gerstmeier, S. Libreros, X. De la Rosa, M. Werner, P.C. Norris, \n",
      "N. Chiang, C.N. Serhan, Human macrophages differentially produce specific \n",
      "resolvin or leukotriene signals that depend on bacterial pathogenicity, Nat. \n",
      "Commun. 9 (1) (2018) 59 . \n",
      "[19] P.M. Jordan, J. Gerstmeier, S. Pace, R. Bilancia, Z. Rao, F. Borner, L. Miek, \n",
      "O. Gutierrez-Gutierrez, V. Arakandy, A. Rossi, A. Ialenti, C. Gonzalez-Estevez, \n",
      "B. Loffler, L. Tuchscherr, C.N. Serhan, O. Werz, Staphylococcus aureus-derived \n",
      "alpha-hemolysin evokes generation of specialized pro-resolving mediators \n",
      "promoting inflammation resolution, Cell Rep. 33 (2) (2020), 108247 . \n",
      "[20] T. Liu, A. Xiang, T. Peng, A.C. Doran, K.J. Tracey, B.J. Barnes, I. Tabas, M. Son, \n",
      "B. Diamond, HMGB1-C1q complexes regulate macrophage function by switching \n",
      "between leukotriene and specialized proresolving mediator biosynthesis, Proc. \n",
      "Natl. Acad. Sci. U. S. A. 116 (46) (2019) 23254 – 23263 . \n",
      "[21] C.N. Serhan, C.B. Clish, J. Brannon, S.P. Colgan, N. Chiang, K. Gronert, Novel \n",
      "functional sets of lipid-derived mediators with antiinflammatory actions \n",
      "generated from omega-3 fatty acids via cyclooxygenase 2-nonsteroidal \n",
      "antiinflammatory drugs and transcellular processing, J. Exp. Med. 192 (2000) \n",
      "1197 – 1204 . \n",
      "[22] N. Chiang, C.N. Serhan, Cell-cell interaction in the transcellular biosynthesis of \n",
      "novel omega-3-derived lipid mediators, Methods Mol. Biol. 341 (2006) 227 – 250 . \n",
      "[23] A.S. Archambault, Y. Zaid, V. Rakotoarivelo, C. Turcotte, \n",
      "´\n",
      "E. Dor ´e, I. Dubuc, \n",
      "C. Martin, O. Flamand, Y. Amar, A. Cheikh, H. Fares, A. El Hassani, Y. Tijani, \n",
      "A. C ˆot ´e, M. Laviolette, \n",
      "´\n",
      "E. Boilard, L. Flamand, N. Flamand, High levels of \n",
      "eicosanoids and docosanoids in the lungs of intubated COVID-19 patients, FASEB \n",
      "J. 35 (6) (2021), e21666 . \n",
      "[24] B. Schwarz, L. Sharma, L. Roberts, X. Peng, S. Bermejo, I. Leighton, A.C. Massana, \n",
      "S. Farhadian, A. Ko, C. DelaCruz, C.M. Bosio, Severe SARS-CoV-2 infection in \n",
      "humans is defined by a shift in the serum lipidome resulting in dysregulation of \n",
      "eicosanoid immune mediators, medRxiv (2020), https://doi.org/10.1101/ \n",
      "2020.07.09.20149849 . Preprint, July 13. \n",
      "[25] D. Koenis, I. Beegun, C. Jouvene, G.A. Aguirre, P.R. Souza, M. Gonzalez-Nunez, \n",
      "L. Ly, K. Pistorius, H.M. Kocher, W. Ricketts, G. Thomas, M. Perretti, G. Alusi, \n",
      "P. Pfeffer, J. Dalli, Disrupted resolution mechanisms favor altered phagocyte \n",
      "responses in Covid-19, Circ. Res. 129 (2021) e54 – e71 . \n",
      "[26] J. Turnbull, R. Jha, C.A. Ortori, E. Lunt, P.J. Tighe, W.L. Irving, S.A. Gohir, D. \n",
      "H. Kim, A.M. Valdes, A.W. Tarr, D.A. Barrett, V. Chapman, Serum levels of pro- \n",
      "inflammatory lipid mediators and specialised pro-resolving molecules are \n",
      "increased in SARS-CoV-2 patients and correlate with markers of the adaptive \n",
      "immune response, J. Infect. Dis. (2022), https://doi.org/10.1093/infdis/jiab632 . \n",
      "[27] R. Yang, N. Chiang, S.F. Oh, C.N. Serhan, Metabolomics-lipidomics of eicosanoids \n",
      "and docosanoids generated by phagocytes, Curr. Protoc. Immunol. Suppl 95 \n",
      "(2011) 14.26.1 – 14.26.26 . \n",
      "[28] C.N. Serhan, Y. Lu, S. Hong, R. Yang, Mediator lipidomics: search algorithms for \n",
      "eicosanoids, resolvins and protectins, Meth. Enzymol. 432 (2007) 275 – 317 . \n",
      "[29] Y. Lu, S. Hong, R. Yang, J. Uddin, K.H. Gotlinger, N.A. Petasis, C.N. Serhan, \n",
      "Identification of endogenous resolvin E1 and other lipid mediators derived from \n",
      "eicosapentaenoic acid via electrospray low energy tandem mass spectrometry: \n",
      "spectra and fragmentation mechanisms, Rapid Commun. Mass Spectrom. 21 \n",
      "(2007) 7 – 22 . \n",
      "[30] S. Hong, Y. Lu, R. Yang, K.H. Gotlinger, N.A. Petasis, C.N. Serhan, Resolvin D1, \n",
      "protectin D1, and related docosahexaenoic acid-derived products: analysis via \n",
      "electrospray/low energy tandem mass spectrometry based on spectra and \n",
      "fragmentation mechanisms, J. Am. Soc. Mass Spectrom. 18 (2007) 128 – 144 . \n",
      "[31] B. Schwarz, L. Sharma, L. Roberts, X. Peng, S. Bermejo, I. Leighton, A. Casanovas- \n",
      "Massana, M. Minasyan, S. Farhadian, A.I. Ko, C.S. Dela Cruz, C.M. Bosio, Cutting \n",
      "edge: severe SARS-CoV-2 infection in humans is defined by a shift in the serum \n",
      "lipidome, resulting in dysregulation of eicosanoid immune mediators, \n",
      "J. Immunol. 206 (2) (2021) 329 – 334 . \n",
      "[32] E. Andreakos, M. Papadaki, C.N. Serhan, Dexamethasone, pro-resolving lipid \n",
      "mediators and resolution of inflammation in COVID-19, Allergy 76 (3) (2021) \n",
      "626 – 628 . \n",
      "[33] D. Panigrahy, M.M. Gilligan, S. Huang, A. Gartung, I. Cort ´es-Puch, P.J. Sime, R. \n",
      "P. Phipps, C.N. Serhan, B.D. Hammock, Inflammation resolution: a dual-pronged \n",
      "approach to averting cytokine storms in COVID-19? Cancer Metastasis Rev. 39 \n",
      "(2020) 337 – 340 . \n",
      "[34] A. Recchiuti, S. Patruno, D. Mattoscio, E. Isopi, A. Pomilio, A. Lamolinara, \n",
      "M. Iezzi, R. Pecce, M. Romano, Resolvin D1 and D2 reduce SARS-CoV-2-induced \n",
      "inflammatory responses in cystic fibrosis macrophages, FASEB J. 35 (4) (2021), \n",
      "e21441 . \n",
      "[35] A.E. Barden, E. Mas, K.D. Croft, M. Phillips, T.A. Mori, Specialized proresolving \n",
      "lipid mediators in humans with the metabolic syndrome after n-3 fatty acids and \n",
      "aspirin, Am. J. Clin. Nutr. 102 (6) (2015) 1357 – 1364 . \n",
      "[36] H.A. Bazan, Y. Lu, B. Jun, Z. Fang, T.C. Woods, S. Hong, Circulating \n",
      "inflammation-resolving lipid mediators RvD1 and DHA are decreased in patients \n",
      "with acutely symptomatic carotid disease, Prostaglandins Leukot. Essent. Fatty \n",
      "Acids 125 (2017) 43 – 47 . \n",
      "[37] A.E. Barden, S. Shinde, V. Burke, I.B. Puddey, L.J. Beilin, A.B. Irish, G.F. Watts, T. \n",
      "A. Mori, The effect of n-3 fatty acids and coenzyme Q10 supplementation on \n",
      "neutrophil leukotrienes, mediators of inflammation resolution and \n",
      "myeloperoxidase in chronic kidney disease, Prostaglandins Other Lipid Mediat. \n",
      "136 (2018) 1 – 8 . \n",
      "[38] I. Hartling, A. Cremonesi, E. Osuna, P.H. Lou, E. Lucchinetti, M. Zaugg, \n",
      "M. Hersberger, Quantitative profiling of inflammatory and pro-resolving lipid \n",
      "mediators in human adolescents and mouse plasma using UHPLC-MS/MS, Clin. \n",
      "Chem. Lab. Med. 59 (2021) 1811 – 1823 . \n",
      "[39] E. Mas, A. Barden, V. Burke, L.J. Beilin, G.F. Watts, R.C. Huang, I.B. Puddey, A. \n",
      "B. Irish, T.A. Mori, A randomized controlled trial of the effects of n-3 fatty acids \n",
      "on resolvins in chronic kidney disease, Clin. Nutr. (Edinburgh, Scotland) 35 (2) \n",
      "(2016) 331 – 336 . \n",
      "[40] E. Mas, K.D. Croft, P. Zahra, A. Barden, T.A. Mori, Resolvins D.1, D2, and other \n",
      "mediators of self-limited resolution of inflammation in human blood following n- \n",
      "3 fatty acid supplementation, Clin. Chem. 58 (2012) 1476 – 1484 . \n",
      "C.N. Serhan et al.                                                                                                                                                                                                                               \n",
      "Seminars in Immunology 59 (2022) 101597\n",
      "14\n",
      "[41] K.J. Polinski, M. Armstrong, J. Manke, J. Seifert, T. Crume, F. Yang, M. Clare- \n",
      "Salzler, V.M. Holers, N. Reisdorph, J.M. Norris, Collection and storage of human \n",
      "plasma for measurement of oxylipins, Metabolites 11 (3) (2021) 137 . \n",
      "[42] F. Schulte, A.A. Asbeutah, P.N. Benotti, G.C. Wood, C. Still, B.R. Bistrian, \n",
      "M. Hardt, F.K. Welty, The relationship between specialized pro-resolving lipid \n",
      "mediators, morbid obesity and weight loss after bariatric surgery, Sci. Rep. 10 (1) \n",
      "(2020) 20128 . \n",
      "[43] A. Barden, E. Mas, K.D. Croft, M. Phillips, T.A. Mori, Short-term n-3 fatty acid \n",
      "supplementation but not aspirin increases plasma proresolving mediators of \n",
      "inflammation, J. Lipid Res. 55 (2014) 2401 – 2407 . \n",
      "[44] G.A. Weiss, H. Troxler, G. Klinke, D. Rogler, C. Braegger, M. Hersberger, High \n",
      "levels of anti-inflammatory and pro-resolving lipid mediators lipoxins and \n",
      "resolvins and declining docosahexaenoic acid levels in human milk during the \n",
      "first month of lactation, Lipids Health Dis. 12 (2013) 89 . \n",
      "[45] A. Barden, S. Shinde, I.J. Tsai, K.D. Croft, L.J. Beilin, I.B. Puddey, T.A. Mori, Effect \n",
      "of weight loss on neutrophil resolvins in the metabolic syndrome, Prostaglandins \n",
      "Leukot. Essent. Fatty Acids 148 (2019) 25 – 29 . \n",
      "[46] A.E. Barden, M. Moghaddami, E. Mas, M. Phillips, L.G. Cleland, T.A. Mori, \n",
      "Specialised pro-resolving mediators of inflammation in inflammatory arthritis, \n",
      "Prostaglandins Leukot. Essent. Fatty Acids 107 (2016) 24 – 29 . \n",
      "[47] J.A. Keelan, E. Mas, N. D ’ Vaz, J.A. Dunstan, S. Li, A.E. Barden, P.J. Mark, B. \n",
      "J. Waddell, S.L. Prescott, T.A. Mori, Effects of maternal n-3 fatty acid \n",
      "supplementation on placental cytokines, pro-resolving lipid mediators and their \n",
      "precursors, Reproduction 149 (2) (2015) 171 – 178 . \n",
      "[48] J.F. Markworth, L.A. Brown, E. Lim, J.A. Castor-Macias, J. Larouche, P.C. \n",
      "D. Macpherson, C. Davis, C.A. Aguilar, K.R. Maddipati, S.V. Brooks, \n",
      "Metabolipidomic profiling reveals an age-related deficiency of skeletal muscle \n",
      "pro-resolving mediators that contributes to maladaptive tissue remodeling, Aging \n",
      "Cell 20 (6) (2021), e13393 . \n",
      "[49] J.F. Markworth, K.B. Sugg, D.C. Sarver, K.R. Maddipati, S.V. Brooks, Local shifts \n",
      "in inflammatory and resolving lipid mediators in response to tendon overuse, \n",
      "FASEB J. 35 (6) (2021), e21655 . \n",
      "[50] S. Lamon-Fava, J. So, D. Mischoulon, T.R. Ziegler, B.W. Dunlop, B. Kinkead, P. \n",
      "J. Schettler, A.A. Nierenberg, J.C. Felger, K.R. Maddipati, M. Fava, M. \n",
      "H. Rapaport, Dose- and time-dependent increase in circulating anti-inflammatory \n",
      "and pro-resolving lipid mediators following eicosapentaenoic acid \n",
      "supplementation in patients with major depressive disorder and chronic \n",
      "inflammation, Prostaglandins Leukot. Essent. Fatty Acids 164 (2021), 102219 . \n",
      "[51] J.F. Markworth, L.D. Vella, B.S. Lingard, D.L. Tull, T.W. Rupasinghe, A.J. Sinclair, \n",
      "K.R. Maddipati, D. Cameron-Smith, Human inflammatory and resolving lipid \n",
      "mediator responses to resistance exercise and ibuprofen treatment, Am. J. \n",
      "Physiol. Regul. Integr. Compar. Physiol. 305 (2013) R1281 – R1296 . \n",
      "[52] N. Terrando, J.J. Park, M. Devinney, C. Chan, M. Cooter, P. Avasarala, J. \n",
      "P. Mathew, Q.J. Quinones, K.R. Maddipati, M. Berger, Immunomodulatory lipid \n",
      "mediator profiling of cerebrospinal fluid following surgery in older adults, Sci. \n",
      "Rep. 11 (1) (2021) 3047 . \n",
      "[53] O. Eickmeier, D. Fussbroich, K. Mueller, F. Serve, C. Smaczny, S. Zielen, \n",
      "R. Schubert, Pro-resolving lipid mediator Resolvin D1 serves as a marker of lung \n",
      "disease in cystic fibrosis, PLoS One 12 (2) (2017), e0171249 . \n",
      "[54] A. Barden, S. Shinde, M. Phillips, L. Beilin, E. Mas, J.M. Hodgson, I. Puddey, T. \n",
      "A. Mori, The effects of alcohol on plasma lipid mediators of inflammation \n",
      "resolution in patients with Type 2 diabetes mellitus, Prostaglandins Leukot. \n",
      "Essent. Fatty Acids 133 (2018) 29 – 34 . \n",
      "[55] V.H.L. See, E. Mas, S.L. Prescott, L.J. Beilin, S. Burrows, A.E. Barden, R.C. Huang, \n",
      "T.A. Mori, Effects of prenatal n-3 fatty acid supplementation on offspring \n",
      "resolvins at birth and 12 years of age: a double-blind, randomised controlled \n",
      "clinical trial, Br. J. Nutr. 118 (2017) 971 – 980 . \n",
      "[56] T.V. Hansen, A. Vik, C.N. Serhan, The protectin family of specialized pro- \n",
      "resolving mediators: potent immunoresolvents enabling innovative approaches to \n",
      "target obesity and diabetes, Front. Pharmacol. 9 (2019) 1582 . \n",
      "[57] A.R. Rodriguez, B.W. Spur, First total syntheses of the pro-resolving lipid \n",
      "mediators 7(S),13(R),20(S)-Resolvin T1 and 7(S),13(R)-Resolvin T4, Tetrahedron \n",
      "Lett. 61 (2020), 151473 . \n",
      "[58] A.R. Rodriguez, B.W. Spur, First total synthesis of the pro-resolving lipid mediator \n",
      "7(S),12(R),13(S)-Resolvin T2 and its 13(R)-epimer, Tetrahedron Lett. 61 (20) \n",
      "(2020), 151857 . \n",
      "[59] N. Chiang, M. Sakuma, A.R. Rodriguez, B.W. Spur, D. Irimia, C.N. Serhan, \n",
      "Resolvin T-series reduce neutrophil extracellular traps, Blood (2021), https://doi. \n",
      "org/10.1182/blood.2021013422 . \n",
      "[60] S.K. Orr, R.A. Colas, J. Dalli, N. Chiang, C.N. Serhan, Proresolving actions of a \n",
      "new resolvin D1 analog mimetic qualifies as an immunoresolvent, Am. J. Physiol. \n",
      "Lung Cell Mol. Physiol. 308 (9) (2015) L904 – 11 . \n",
      "[61] S. Libreros, A.E. Shay, R. Nshimiyimana, D. Fichtner, M.J. Martin, N. Wourms, C. \n",
      "N. Serhan, A new E-series resolvin: RvE4 stereochemistry and function in \n",
      "efferocytosis of inflammation-resolution, Front. Immunol. 11 (2020), 631319 . \n",
      "[62] M.L. Edin, D.C. Zeldin, Commercial scale production of RvD4 opens the resolving \n",
      "door to new research, J. Leukoc. Biol. (2018), https://doi.org/10.1002/ \n",
      "jlb.3ce0118-032r . \n",
      "[63] H. Hasturk, F. Schulte, M. Martins, H. Sherzai, C. Floros, M. Cugini, C.J. Chiu, \n",
      "M. Hardt, T. Van Dyke, Safety and preliminary efficacy of a novel host- \n",
      "modulatory therapy for reducing gingival inflammation, Front. Immunol. 12 \n",
      "(2021), 704163 . \n",
      "[64] E. Brennan, P. Kantharidis, M.E. Cooper, C. Godson, Pro-resolving lipid \n",
      "mediators: regulators of inflammation, metabolism and kidney function, Nat. \n",
      "Rev. Nephrol. 17 (11) (2021) 725 – 739 . \n",
      "[65] T. Vartak, C. Godson, E. Brennan, Therapeutic potential of pro-resolving \n",
      "mediators in diabetic kidney disease, Adv. Drug Deliv. Rev. 178 (2021), 113965 . \n",
      "[66] T.W. Vickery, M. Armstrong, J.M. Kofonow, C.E. Robertson, M.E. Kroehl, N. \n",
      "A. Reisdorph, V.R. Ramakrishnan, D.N. Frank, Altered tissue specialized pro- \n",
      "resolving mediators in chronic rhinosinusitis, Prostaglandins Leukot. Essent. Fatty \n",
      "Acids 164 (2021), 102218 . \n",
      "[67] C.A. P ´erez-Novo, J.B. Watelet, C. Claeys, P. Van Cauwenberge, C. Bachert, \n",
      "Prostaglandin, leukotriene, and lipoxin balance in chronic rhinosinusitis with and \n",
      "without nasal polyposis, J. Allergy Clin. Immunol. 115 (6) (2005) 1189 – 1196 . \n",
      "[68] R. De Caterina, n-3 fatty acids in cardiovascular disease, N. Engl. J. Med. 364 (25) \n",
      "(2011) 2439 – 2450 . \n",
      "[69] W.S. Harris, P.C. Calder, D. Mozaffarian, C.N. Serhan, Bang and Dyerberg ’ s \n",
      "omega-3 discovery turns fifty, Nat. Food 2 (2021) 303 – 305 . \n",
      "[70] D.L. Bhatt, P.G. Steg, M. Miller, E.A. Brinton, T.A. Jacobson, S.B. Ketchum, R. \n",
      "T. Doyle Jr., R.A. Juliano, L. Jiao, C. Granowitz, J.C. Tardif, C.M. Ballantyne, R.- \n",
      "I. Investigators, Cardiovascular risk reduction with icosapent ethyl for \n",
      "hypertriglyceridemia, N. Engl. J. Med. 380 (1) (2019) 11 – 22 . \n",
      "[71] A.M. Darwesh, W. Bassiouni, D.K. Sosnowski, J.M. Seubert, Can N-3 \n",
      "polyunsaturated fatty acids be considered a potential adjuvant therapy for \n",
      "COVID-19-associated cardiovascular complications? Pharmacol. Ther. 219 \n",
      "(2021), 107703 . \n",
      "[72] M. Arita, F. Bianchini, J. Aliberti, A. Sher, N. Chiang, S. Hong, R. Yang, N. \n",
      "A. Petasis, C.N. Serhan, Stereochemical assignment, anti-inflammatory \n",
      "properties, and receptor for the omega-3 lipid mediator resolvin E1, J. Exp. Med. \n",
      "201 (2005) 713 – 722 . \n",
      "[73] R.S. Cotran, V. Kumar, T. Collins, Robbins Pathologic Basis of Disease, W.B. \n",
      "Saunders Co., Philadelphia, 1999 . \n",
      "[74] S.P. Colgan, Cell-cell interactions: methods and protocols, in: J.M. Walker (Ed.), \n",
      "Methods in Molecular Biology, Humana Press, Totowa, NJ, 2006 . \n",
      "[75] M. Dona, G. Fredman, J.M. Schwab, N. Chiang, M. Arita, A. Goodarzi, G. Cheng, \n",
      "U.H. von Andrian, C.N. Serhan, Resolvin E1, an EPA-derived mediator in whole \n",
      "blood, selectively counterregulates leukocytes and platelets, Blood 112 (2008) \n",
      "848 – 855 . \n",
      "[76] M. Arita, C.B. Clish, C.N. Serhan, The contributions of aspirin and microbial \n",
      "oxygenase to the biosynthesis of anti-inflammatory resolvins: novel oxygenase \n",
      "products from omega-3 polyunsaturated fatty acids, Biochem. Biophys. Res. \n",
      "Commun. 338 (1) (2005) 149 – 157 . \n",
      "[77] S.F. Oh, P.S. Pillai, A. Recchiuti, R. Yang, C.N. Serhan, Pro-resolving actions and \n",
      "stereoselective biosynthesis of 18 S E-series resolvins in human leukocytes and \n",
      "murine inflammation, J. Clin. Invest. 121 (2011) 569 – 581 . \n",
      "[78] E. Tjonahen, S.F. Oh, J. Siegelman, S. Elangovan, K.B. Percarpio, S. Hong, \n",
      "M. Arita, C.N. Serhan, Resolvin E2: Identification and anti-inflammatory actions: \n",
      "pivotal role of human 5-lipoxygenase in resolvin E series biosynthesis, Chem. \n",
      "Biol. 13 (2006) 1193 – 1202 . \n",
      "[79] C. Trilleaud, V. Gauttier, K. Biteau, I. Girault, L. Belarif, C. Mary, S. Pengam, \n",
      "G. Teppaz, V. Thepenier, R. Danger, G. Robert-Siegwald, M. N ´eel, S. Bruneau, \n",
      "A. Gl ´emain, A. N ´eel, A. Poupon, J.F. Mosnier, G. Ch ˆene, M. Dubourdeau, \n",
      "G. Blancho, B. Vanhove, N. Poirier, Agonist anti-ChemR23 mAb reduces tissue \n",
      "neutrophil accumulation and triggers chronic inflammation resolution, Sci. Adv. \n",
      "7 (14) (2021), eabd1453 . \n",
      "[80] J.M. Schwab, N. Chiang, M. Arita, C.N. Serhan, Resolvin E1 and protectin D1 \n",
      "activate inflammation-resolution programmes, Nature 447 (2007) 869 – 874 . \n",
      "[81] H. Hasturk, A. Kantarci, T. Ohira, M. Arita, N. Ebrahimi, N. Chiang, N.A. Petasis, \n",
      "B.D. Levy, C.N. Serhan, T.E. Van Dyke, RvE1 protects from local inflammation \n",
      "and osteoclast mediated bone destruction in periodontitis, FASEB J. 20 (2006) \n",
      "401 – 403 . \n",
      "[82] M. Arita, S. Oh, T. Chonan, S. Hong, S. Elangovan, Y.-P. Sun, J. Uddin, N. \n",
      "A. Petasis, C.N. Serhan, Metabolic inactivation of resolvin E1 and stabilization of \n",
      "its anti-inflammatory actions, J. Biol. Chem. 281 (2006) 22847 – 22854 . \n",
      "[83] Y. Isobe, M. Arita, S. Matsueda, R. Iwamoto, T. Fujihara, H. Nakanishi, \n",
      "R. Taguchi, K. Masuda, K. Sasaki, D. Urabe, M. Inoue, H. Arai, Identification and \n",
      "structure determination of novel anti-inflammatory mediator resolvin E3, 17,18- \n",
      "dihydroxyeicosapentaenoic acid, J. Biol. Chem. 287 (13) (2012) 10525 – 10534 . \n",
      "[84] J. Endo, M. Sano, Y. Isobe, K. Fukuda, J.X. Kang, H. Arai, M. Arita, 18-HEPE, an n- \n",
      "3 fatty acid metabolite released by macrophages, prevents pressure overload- \n",
      "induced maladaptive cardiac remodeling, J. Exp. Med. 211 (8) (2014) \n",
      "1673 – 1687 . \n",
      "[85] P.C. Norris, S. Libreros, C.N. Serhan, Resolution metabolomes activated by \n",
      "hypoxic environment, Sci. Adv. 5 (2019), eaax4895 . \n",
      "[86] L. Kutzner, K. Goloshchapova, K.M. Rund, M. Jubermann, M. Blum, M. Rothe, S. \n",
      "F. Kirsch, W.H. Schunck, H. Kuhn, N.H. Schebb, Human lipoxygenase isoforms \n",
      "form complex patterns of double and triple oxygenated compounds from \n",
      "eicosapentaenoic acid, Biochim. Biophys. Acta Mol. Cell Biol. Lipids 1865 (12) \n",
      "(2020), 158806 . \n",
      "[87] A.F. Reinertsen, K.G. Primdahl, A.E. Shay, C.N. Serhan, T.V. Hansen, M. Aursnes, \n",
      "Stereoselective synthesis and structural confirmation of the specialized pro- \n",
      "resolving mediator resolvin E4, J. Org. Chem. 86 (4) (2021) 3535 – 3545 . \n",
      "[88] M.A. Cassatella, The Neutrophil an Emergenging Regulator of Inflammatory and \n",
      "Immune Response, 2003 . \n",
      "[89] W.M. Nauseef, N. Borregaard, Neutrophils at work, Nat. Immunol. 15 (7) (2014) \n",
      "602 – 611 . \n",
      "[90] K.V. Do, E. Hjorth, Y. Wang, B. Jun, M.A. Kautzmann, M. Eriksdotter, \n",
      "M. Schultzberg, N.G. Bazan, Cerebrospinal fluid profile of lipid mediators in \n",
      "Alzheimer ’ s disease, Res. Sq. (2021), https://doi.org/10.21203/rs.3.rs-1029034/ \n",
      "v1 [Preprint]. \n",
      "C.N. Serhan et al.                                                                                                                                                                                                                               \n",
      "Seminars in Immunology 59 (2022) 101597\n",
      "15\n",
      "[91] Y. Okabe, R. Medzhitov, Tissue biology perspective on macrophages, Nat. \n",
      "Immunol. 17 (1) (2016) 9 – 17 . \n",
      "[92] J. Lee, J.U. An, T.H. Kim, Y.J. Ko, J.B. Park, D.K. Oh, Discovery and engineering \n",
      "of a microbial double-oxygenating lipoxygenase for synthesis of dihydroxy fatty \n",
      "acids as specialized proresolving mediators, ACS Sustain. Chem. Eng. 8 (2020) \n",
      "16172 – 16183 . \n",
      "[93] J. Cl `aria, C.N. Serhan, Aspirin triggers previously undescribed bioactive \n",
      "eicosanoids by human endothelial cell-leukocyte interactions, Proc. Natl. Acad. \n",
      "Sci. U. S. A. 92 (1995) 9475 – 9479 . \n",
      "[94] J.Y. Lee, S.H. Han, M.H. Park, I.S. Song, M.K. Choi, E. Yu, C.M. Park, H.J. Kim, S. \n",
      "H. Kim, E.H. Schuchman, H.K. Jin, J.S. Bae, N-AS-triggered SPMs are direct \n",
      "regulators of microglia in a model of Alzheimer ’ s disease, Nat. Commun. 11 (1) \n",
      "(2020) 2358 . \n",
      "[95] J. Dalli, N. Chiang, C.N. Serhan, Elucidation of novel 13-series resolvins that \n",
      "increase with atorvastatin and clear infections, Nat. Med. 21 (9) (2015) \n",
      "1071 – 1075 . \n",
      "[96] O. Radmark, O. Werz, D. Steinhilber, B. Samuelsson, 5-Lipoxygenase, a key \n",
      "enzyme for leukotriene biosynthesis in health and disease, Biochim. Biophys. Acta \n",
      "1851 (4) (2015) 331 – 339 . \n",
      "[97] Z. He, D. Tao, J. Xiong, F. Lou, J. Zhang, J. Chen, W. Dai, J. Sun, Y. Wang, \n",
      "Phosphorylation of 5-LOX: the potential set-point of inflammation, Neurochem. \n",
      "Res. 45 (10) (2020) 2245 – 2257 . \n",
      "[98] Y. Ye, Y. Lin, J.R. Perez-Polo, B.F. Uretsky, Z. Ye, B.C. Tieu, Y. Birnbaum, \n",
      "Phosphorylation of 5-lipoxygenase at ser523 by protein kinase A determines \n",
      "whether pioglitazone and atorvastatin induce proinflammatory leukotriene B4 or \n",
      "anti-inflammatory 15-epi-lipoxin a4 production, J. Immunol. 181 (5) (2008) \n",
      "3515 – 3523 . \n",
      "[99] M. Arita, M. Yoshida, S. Hong, E. Tjonahen, J.N. Glickman, N.A. Petasis, R. \n",
      "S. Blumberg, C.N. Serhan, Resolvin E1, an endogenous lipid mediator derived \n",
      "from omega-3 eicosapentaenoic acid, protects against 2,4,6-trinitrobenzene \n",
      "sulfonic acid-induced colitis, Proc. Natl. Acad. Sci. U. S. A. 102 (21) (2005) \n",
      "7671 – 7676 . \n",
      "[100] Y. Sawada, T. Honda, S. Hanakawa, S. Nakamizo, T. Murata, Y. Ueharaguchi- \n",
      "Tanada, S. Ono, W. Amano, S. Nakajima, G. Egawa, H. Tanizaki, A. Otsuka, \n",
      "A. Kitoh, T. Dainichi, N. Ogawa, Y. Kobayashi, T. Yokomizo, M. Arita, \n",
      "M. Nakamura, Y. Miyachi, K. Kabashima, Resolvin E1 inhibits dendritic cell \n",
      "migration in the skin and attenuates contact hypersensitivity responses, J. Exp. \n",
      "Med. 212 (11) (2015) 1921 – 1930 . \n",
      "[101] C.T. Lee, R. Teles, A. Kantarci, T. Chen, J. McCafferty, J.R. Starr, L.C. Brito, B. \n",
      "J. Paster, T.E. Van Dyke, Resolvin E1 reverses experimental periodontitis and \n",
      "dysbiosis, J. Immunol. 197 (7) (2016) 2796 – 2806 . \n",
      "[102] C.K. Park, Z.Z. Xu, T. Liu, N. Lu, C.N. Serhan, R.R. Ji, Resolvin D2 is a potent \n",
      "endogenous inhibitor for transient receptor potential subtype V1/A1, \n",
      "inflammatory pain, and spinal cord synaptic plasticity in mice: distinct roles of \n",
      "resolvin D1, D2, and E1, J. Neurosci. 31 (50) (2011) 18433 – 18438 . \n",
      "[103] E. Albuquerque-Souza, F. Schulte, T. Chen, M. Hardt, H. Hasturk, T.E. Van Dyke, \n",
      "M. Holzhausen, A. Kantarci, Maresin-1 and resolvin E1 promote regenerative \n",
      "properties of periodontal ligament stem cells under inflammatory conditions, \n",
      "Front. Immunol. 11 (2020), 585530 . \n",
      "[104] K.M. Reddoch-Cardenas, U. Sharma, C.L. Salgado, C. Cantu, D.N. Darlington, H. \n",
      "F. Pidcoke, J.A. Bynum, A.P. Cap, Use of specialized pro-resolving mediators to \n",
      "alleviate cold platelet storage lesion, Transfusion 60 (Suppl. 3) (2020) S112 – s118 . \n",
      "[105] J.M. Thornton, K. Yin, Role of specialized pro-resolving mediators in modifying \n",
      "host defense and decreasing bacterial virulence, Molecules 26 (22) (2021) 6970 . \n",
      "[106] J. Chen, G.S.D. Purvis, D. Collotta, S. Al Zoubi, M.A. Sugimoto, A. Cacace, \n",
      "L. Martin, R.A. Colas, M. Collino, J. Dalli, C. Thiemermann, RvE1 attenuates \n",
      "polymicrobial sepsis-induced cardiac dysfunction and enhances bacterial \n",
      "clearance, Front. Immunol. 11 (2020) 2080 . \n",
      "[107] K.T. Keyes, Y. Ye, Y. Lin, C. Zhang, J.R. Perez-Polo, P. Gjorstrup, Y. Birnbaum, \n",
      "Resolvin E1 protects the rat heart against reperfusion injury, Am. J. Physiol. Heart \n",
      "Circ. Physiol. 299 (1) (2010) H153 – 64 . \n",
      "[108] M.O. Freire, J. Dalli, C.N. Serhan, T.E. Van Dyke, Neutrophil resolvin E1 receptor \n",
      "expression and function in type 2 diabetes, J. Immunol. 198 (2) (2017) 718 – 728 . \n",
      "[109] E.L. Campbell, C.F. MacManus, D.J. Kominsky, S. Keely, L.E. Glover, B.E. Bowers, \n",
      "M. Scully, W.J. Bruyninckx, S.P. Colgan, Resolvin E1-induced intestinal alkaline \n",
      "phosphatase promotes resolution of inflammation through LPS detoxification, \n",
      "Proc. Natl. Acad. Sci. U. S. A. 107 (32) (2010) 14298 – 14303 . \n",
      "[110] M. Quiros, D. Feier, D. Birkl, R. Agarwal, D.W. Zhou, A.J. Garcia, C.A. Parkos, \n",
      "A. Nusrat, Resolvin E1 is a pro-repair molecule that promotes intestinal epithelial \n",
      "wound healing, Proc. Natl. Acad. Sci. U. S. A. 117 (17) (2020) 9477 – 9482 . \n",
      "[111] O. Haworth, M. Cernadas, B.D. Levy, NK cells are effectors for resolvin E1 in the \n",
      "timely resolution of allergic airway inflammation, J. Immunol. 186 (11) (2011) \n",
      "6129 – 6135 . \n",
      "[112] A. Gonzalez-Periz, R. Horrillo, N. Ferre, K. Gronert, B. Dong, E. Moran-Salvador, \n",
      "E. Titos, M. Martinez-Clemente, M. Lopez-Parra, V. Arroyo, J. Claria, Obesity- \n",
      "induced insulin resistance and hepatic steatosis are alleviated by omega-3 fatty \n",
      "acids: a role for resolvins and protectins, FASEB J. 23 (6) (2009) 1946 – 1957 . \n",
      "[113] K.J. Ho, M. Spite, C.D. Owens, H. Lancero, A.H. Kroemer, R. Pande, M.A. Creager, \n",
      "C.N. Serhan, M.S. Conte, Aspirin-triggered lipoxin and resolvin E1 modulate \n",
      "vascular smooth muscle phenotype and correlate with peripheral atherosclerosis, \n",
      "Am. J. Pathol. 177 (4) (2010) 2116 – 2123 . \n",
      "[114] M.L. Sulciner, C.N. Serhan, M.M. Gilligan, D.K. Mudge, J. Chang, A. Gartung, K. \n",
      "A. Lehner, D.R. Bielenberg, B. Schmidt, J. Dalli, E.R. Greene, Y. Gus-Brautbar, \n",
      "J. Piwowarski, T. Mammoto, D. Zurakowski, M. Perretti, V.P. Sukhatme, \n",
      "A. Kaipainen, M.W. Kieran, S. Huang, D. Panigrahy, Resolvins suppress tumor \n",
      "growth and enhance cancer therapy, J. Exp. Med. 215 (1) (2018) 115 – 140 . \n",
      "[115] Y. Sawada, T. Honda, S. Nakamizo, A. Otsuka, N. Ogawa, Y. Kobayashi, \n",
      "M. Nakamura, K. Kabashima, Resolvin E1 attenuates murine psoriatic dermatitis, \n",
      "Sci. Rep. 8 (1) (2018) 11873 . \n",
      "[116] E.J. Haas-Stapleton, Y. Lu, S. Hong, M. Arita, S. Favoreto, S. Nigam, C.N. Serhan, \n",
      "N. Agabian, Candida albicans modulates host defense by biosynthesizing the pro- \n",
      "resolving mediator resolvin E1, PLoS One 2 (12) (2007) e1316 . \n",
      "[117] N.K. Rajasagi, P.B. Reddy, A. Suryawanshi, S. Mulik, P. Gjorstrup, B.T. Rouse, \n",
      "Controlling herpes simplex virus-induced ocular inflammatory lesions with the \n",
      "lipid-derived mediator resolvin E1, J. Immunol. 186 (3) (2011) 1735 – 1746 . \n",
      "[118] Z.Z. Xu, L. Zhang, T. Liu, J.Y. Park, T. Berta, R. Yang, C.N. Serhan, R.R. Ji, \n",
      "Resolvins RvE1 and RvD1 attenuate inflammatory pain via central and peripheral \n",
      "actions, Nat. Med. 16 (5) (2010) 592 – 597, 1p following 597 . \n",
      "[119] R.A. Colas, P.R. Souza, M.E. Walker, M. Burton, Z. Zaslona, A.M. Curtis, R. \n",
      "M. Marques, J. Dalli, Impaired production and diurnal regulation of vascular \n",
      "RvDn-3 DPA increase systemic inflammation and cardiovascular disease, Circ. \n",
      "Res. 122 (6) (2018) 855 – 863 . \n",
      "[120] J.L. Ramirez, W.J. Gasper, S.A. Khetani, G.J. Zahner, N.K. Hills, P.T. Mitchell, B. \n",
      "E. Sansbury, M.S. Conte, M. Spite, S.M. Grenon, Fish oil increases specialized pro- \n",
      "resolving lipid mediators in PAD (The OMEGA-PAD II trial), J. Surg. Res. 238 \n",
      "(2019) 164 – 174 . \n",
      "[121] H. Arnardottir, S.K. Orr, J. Dalli, C.N. Serhan, Human milk proresolving \n",
      "mediators stimulate resolution of acute inflammation, Mucosal Immunol. 9 (3) \n",
      "(2016) 757 – 766 . \n",
      "[122] M.P. Motwani, R.A. Colas, M.J. George, J.D. Flint, J. Dalli, A. Richard-Loendt, R. \n",
      "P. Dee Maeyer, C.N. Serhan, D.W. Gilroy, Pro-resolving mediators promote \n",
      "resolution in a human skin model of UV-killed Escherichia coli-driven acute \n",
      "inflammation, JCI Insight 3 (2018), e94463 . \n",
      "[123] S. Deyama, K. Shimoda, H. Suzuki, Y. Ishikawa, K. Ishimura, H. Fukuda, \n",
      "N. Hitora-Imamura, S. Ide, M. Satoh, K. Kaneda, S. Shuto, M. Minami, Resolvin \n",
      "E1/E2 ameliorate lipopolysaccharide-induced depression-like behaviors via \n",
      "ChemR23, Psychopharmacology (Berl.) 235 (1) (2018) 329 – 336 . \n",
      "[124] Y. Murakami, H. Fukuda, R. Muromoto, K. Hirashima, K. Ishimura, K. Fujiwara, \n",
      "J. Ishihara, T. Matsuda, M. Watanabe, S. Shuto, Design and synthesis of benzene \n",
      "congeners of resolvin E2, a proresolving lipid mediator, as its stable equivalents, \n",
      "ACS Med. Chem. Lett. 11 (4) (2020) 479 – 484 . \n",
      "[125] M. Sato, H. Aoki-Saito, H. Fukuda, H. Ikeda, Y. Koga, M. Yatomi, H. Tsurumaki, \n",
      "T. Maeno, T. Saito, T. Nakakura, T. Mori, M. Yanagawa, M. Abe, Y. Sako, \n",
      "K. Dobashi, T. Ishizuka, M. Yamada, S. Shuto, T. Hisada, Resolvin E3 attenuates \n",
      "allergic airway inflammation via the interleukin-23-interleukin-17A pathway, \n",
      "FASEB J. 33 (11) (2019) 12750 – 12759 . \n",
      "[126] S. Deyama, K. Shimoda, H. Ikeda, H. Fukuda, S. Shuto, M. Minami, Resolvin E3 \n",
      "attenuates lipopolysaccharide-induced depression-like behavior in mice, \n",
      "J. Pharmacol. Sci. 138 (1) (2018) 86 – 88 . \n",
      "[127] A. Yamashita, K. Kawana, K. Tomio, A. Taguchi, Y. Isobe, R. Iwamoto, K. Masuda, \n",
      "H. Furuya, T. Nagamatsu, K. Nagasaka, T. Arimoto, K. Oda, O. Wada-Hiraike, \n",
      "T. Yamashita, Y. Taketani, J.X. Kang, S. Kozuma, H. Arai, M. Arita, Y. Osuga, \n",
      "T. Fujii, Increased tissue levels of omega-3 polyunsaturated fatty acids prevents \n",
      "pathological preterm birth, Sci. Rep. 3 (2013) 3113 . \n",
      "[128] H. Fukuda, H. Ikeda, R. Muromoto, K. Hirashima, K. Ishimura, K. Fujiwara, \n",
      "H. Aoki-Saito, T. Hisada, M. Watanabe, J. Ishihara, T. Matsuda, S. Shuto, \n",
      "Synthesis of resolvin E3, a proresolving lipid mediator, and its deoxy derivatives: \n",
      "identification of 18-deoxy-resolvin E3 as a potent anti-inflammatory agent, \n",
      "J. Org. Chem. 85 (21) (2020) 14190 – 14200, https://doi.org/10.1021/acs. \n",
      "joc.0c01701 . \n",
      "[129] S. Deyama, Resolvins as novel targets for rapid-acting antidepressants, Folia \n",
      "Pharmacol. Jpn. 155 (6) (2020) 381 – 385 . \n",
      "[130] D.L. Bhatt, P.G. Steg, M. Miller, Cardiovascular risk reduction with icosapent \n",
      "ethyl. reply, N. Engl. J. Med. 380 (17) (2019) 1678 . \n",
      "[131] A. Tan, B. Sullenbarger, R. Prakash, J.C. McDaniel, Supplementation with \n",
      "eicosapentaenoic acid and docosahexaenoic acid reduces high levels of circulating \n",
      "proinflammatory cytokines in aging adults: a randomized, controlled study, \n",
      "Prostaglandins Leukot. Essent. Fatty Acids 132 (2018) 23 – 29 . \n",
      "[132] P.R. Souza, R.M. Marques, E.A. Gomez, R.A. Colas, R. De Matteis, A. Zak, \n",
      "M. Patel, D.J. Collier, J. Dalli, Enriched marine oil supplements increase \n",
      "peripheral blood specialized pro-resolving mediators concentrations and \n",
      "reprogram host immune responses: a randomized double-blind placebo-controlled \n",
      "study, Circ. Res. 126 (1) (2020) 75 – 90 . \n",
      "[133] F.K. Welty, F. Schulte, A. Alfaddagh, T.K. Elajami, B.R. Bistrian, M. Hardt, \n",
      "Regression of human coronary artery plaque is associated with a high ratio of (18- \n",
      "hydroxy-eicosapentaenoic acid + resolvin E1) to leukotriene B(4), FASEB J. 35 \n",
      "(4) (2021), e21448 . \n",
      "[134] E.C. Keeley, H.J. Li, C.R. Cogle, E.M. Handberg, C.N.B. Merz, C.J. Pepine, \n",
      "Specialized proresolving mediators in symptomatic women with coronary \n",
      "microvascular dysfunction (from the women ’ s ischemia trial to reduce events in \n",
      "nonobstructive CAD [WARRIOR] trial), Am. J. Cardiol. 162 (2022) 1 – 5 . \n",
      "[135] Y.-P. Sun, S.F. Oh, J. Uddin, R. Yang, K. Gotlinger, E. Campbell, S.P. Colgan, N. \n",
      "A. Petasis, C.N. Serhan, Resolvin D1 and its aspirin-triggered 17 R epimer: \n",
      "stereochemical assignments, anti-inflammatory properties and enzymatic \n",
      "inactivation, J. Biol. Chem. 282 (2007) 9323 – 9334 . \n",
      "[136] M. Spite, L.V. Norling, L. Summers, R. Yang, D. Cooper, N.A. Petasis, R.J. Flower, \n",
      "M. Perretti, C.N. Serhan, Resolvin D2 is a potent regulator of leukocytes and \n",
      "controls microbial sepsis, Nature 461 (2009) 1287 – 1291 . \n",
      "[137] J. Dalli, J.W. Winkler, R.A. Colas, H. Arnardottir, C.Y.C. Cheng, N. Chiang, N. \n",
      "A. Petasis, C.N. Serhan, Resolvin D3 and aspirin-triggered resolvin D3 are potent \n",
      "immunoresolvents, Chem. Biol. 20 (2013) 188 – 201 . \n",
      "C.N. Serhan et al.                                                                                                                                                                                                                               \n",
      "Seminars in Immunology 59 (2022) 101597\n",
      "16\n",
      "[138] P.C. Norris, H. Arnardottir, J.M. Sanger, D. Fichtner, G.S. Keyes, C.N. Serhan, \n",
      "Resolvin D3 multi-level proresolving actions are host protective during infection, \n",
      "Prostaglandins Leukot. Essent. Fatty Acids 138 (2018) 81 – 89 . \n",
      "[139] J.W. Winkler, J. Uddin, C.N. Serhan, N.A. Petasis, Stereocontrolled total synthesis \n",
      "of the potent anti-inflammatory and pro-resolving lipid mediator resolvin D3 and \n",
      "its aspirin-triggered 17R-epimer, Org. Lett. 15 (7) (2013) 1424 – 1427 . \n",
      "[140] J.W. Winkler, S. Libreros, X. De La Rosa, B.E. Sansbury, P.C. Norris, N. Chiang, \n",
      "D. Fichtner, G.S. Keyes, N. Wourms, M. Spite, C.N. Serhan, Structural insights into \n",
      "Resolvin D4 actions and further metabolites via a new total organic synthesis and \n",
      "validation, J. Leukoc. Biol. 103 (2018) 995 – 1010 . \n",
      "[141] J.W. Winkler, S.K. Orr, J. Dalli, C.Y. Cheng, J.M. Sanger, N. Chiang, N.A. Petasis, \n",
      "C.N. Serhan, Resolvin D4 stereoassignment and its novel actions in host \n",
      "protection and bacterial clearance, Sci. Rep. 6 (2016) 18972 . \n",
      "[142] A.E. Shay, R. Nshimiyimana, B. Samuelsson, N.A. Petasis, J.Z. Haeggstrom, C. \n",
      "N. Serhan, Human leukocytes selectively convert 4S,5S-epoxy-resolvin to resolvin \n",
      "D3, resolvin D4, and a cys-resolvin isomer, Proc. Natl. Acad. Sci. U. S. A. 118 (51) \n",
      "(2021) . \n",
      "[143] D. Cherpokova, C.C. Jouvene, S. Libreros, E. DeRoo, L. Chu, X. de la Rosa, \n",
      "P. Norris, D.D. Wagner, C.N. Serhan, Resolvin D4 attenuates the severity of \n",
      "pathological thrombosis in mice, Blood 134 (2019) 1458 – 1468 . \n",
      "[144] M. Holinstat, Resolvin the clot: DVT resolution through RvD4, Blood 134 (2019) \n",
      "1370 – 1371 . \n",
      "[145] K. Huynh, Resolution of thrombus burden by targeting inflammatory mediators to \n",
      "treat DVT, Nat. Rev. Cardiol. 16 (10) (2019) 577 . \n",
      "[146] M. Morita, Y. Kobayashi, Stereocontrolled synthesis of resolvin D4, J. Org. Chem. \n",
      "83 (7) (2018) 3906 – 3914 . \n",
      "[147] L.A. Diaz, N.H. Altman, W.N. Khan, C.N. Serhan, B. Adkins, Specialized \n",
      "proresolving mediators rescue infant mice from lethal Citrobacter rodentium \n",
      "infection and promote immunity against reinfection, Infect. Immun. 85 (10) \n",
      "(2017) e00464 – 17 . \n",
      "[148] A.R. Rodriguez, B.W. Spur, First total synthesis of 7(S),17(S)-resolvin D5, a potent \n",
      "anti-inflammatory docosanoid, Tetrahedron Lett. 46 (2005) 3623 – 3627 . \n",
      "[149] N. Ogawa, T. Sugiyama, M. Morita, Y. Suganuma, Y. Kobayashi, Total synthesis of \n",
      "resolvin D5, J. Org. Chem. 82 (4) (2017) 2032 – 2039 . \n",
      "[150] S.C. Perry, C. Kalyanaraman, B.E. Tourdot, W.S. Conrad, O. Akinkugbe, J. \n",
      "C. Freedman, M. Holinstat, M.P. Jacobson, T.R. Holman, 15-Lipoxygenase-1 \n",
      "biosynthesis of 7S,14S-diHDHA implicates 15-lipoxygenase-2 in biosynthesis of \n",
      "resolvin D5, J. Lipid Res. 61 (7) (2020) 1087 – 1103 . \n",
      "[151] A.V. Sorokin, P. Norris, J. English, A.K. Dey, A. Chaturvedi, Y. Baumer, \n",
      "J. Silverman, M.P. Playford, C.N. Serhan, N.N. Mehta, Identification of pro- \n",
      "resolving and inflammatory lipid mediators in human psoriasis, J. Clin. Lipidol. \n",
      "12 (4) (2018) 1047 – 1060 . \n",
      "[152] F. Rich ´e, B.G. Chousterman, P. Valleur, A. Mebazaa, J.M. Launay, E. Gayat, \n",
      "Protracted immune disorders at one year after ICU discharge in patients with \n",
      "septic shock, Crit. Care 22 (2018) 42 . \n",
      "[153] X. Luo, Y. Gu, X. Tao, C.N. Serhan, R.R. Ji, Resolvin D5 inhibits neuropathic and \n",
      "inflammatory pain in male but not female mice: distinct actions of D-series \n",
      "resolvins in chemotherapy-induced peripheral neuropathy, Front. Pharmacol. 10 \n",
      "(2019) 745 . \n",
      "[154] A. Ariel, P.-L. Li, W. Wang, W.-X. Tang, G. Fredman, S. Hong, K.H. Gotlinger, C. \n",
      "N. Serhan, The docosatriene protectin D1 is produced by T\n",
      "H\n",
      "2 skewing and \n",
      "promotes human T cell apoptosis via lipid raft clustering, J. Biol. Chem. 280 \n",
      "(2005) 43079 – 43086 . \n",
      "[155] V. Chiurchiu, A. Leuti, J. Dalli, A. Jacobsson, L. Battistini, M. Maccarrone, C. \n",
      "N. Serhan, Proresolving lipid mediators resolvin D1, resolvin D2, and maresin 1 \n",
      "are critical in modulating T cell responses, Sci. Transl. Med. 8 (353) (2016), \n",
      "353ra111 . \n",
      "[156] S. Ramon, S. Bancos, T.H. Thatcher, T.I. Murant, S. Moshkani, J.M. Sahler, \n",
      "A. Bottaro, P.J. Sime, R.P. Phipps, Peroxisome proliferator-activated receptor \n",
      "gamma B cell-specific-deficient mice have an impaired antibody response, \n",
      "J. Immunol. 189 (10) (2012) 4740 – 4747 . \n",
      "[157] H. Yamada, J. Saegusa, S. Sendo, Y. Ueda, T. Okano, M. Shinohara, A. Morinobu, \n",
      "Effect of resolvin D5 on T cell differentiation and osteoclastogenesis analyzed by \n",
      "lipid mediator profiling in the experimental arthritis, Sci. Rep. 11 (1) (2021) \n",
      "17312 . \n",
      "[158] T.L. Pham, J. He, A.H. Kakazu, J. Calandria, K.V. Do, R. Nshimiyimana, N. \n",
      "A. Petasis, H.E.P. Bazan, N.G. Bazan, Elovanoid-N32 or RvD6-isomer decrease \n",
      "ACE2 and binding of S protein RBD after injury or INF γ in the eye, Res. Sq. \n",
      "(2020), https://doi.org/10.21203/rs.3.rs-55764/v1 [Preprint]. \n",
      "[159] T.L. Pham, J. He, A.H. Kakazu, J. Calandria, K.V. Do, R. Nshimiyimana, T.F. Lam, \n",
      "N.A. Petasis, H.E.P. Bazan, N.G. Bazan, ELV-N32 and RvD6 isomer decrease pro- \n",
      "inflammatory cytokines, senescence programming, ACE2 and SARS-CoV-2-spike \n",
      "protein RBD binding in injured cornea, Sci. Rep. 11 (1) (2021) 12787 . \n",
      "[160] T.L. Pham, H.E.P. Bazan, Docosanoid signaling modulates corneal nerve \n",
      "regeneration: effect on tear secretion, wound healing, and neuropathic pain, \n",
      "J. Lipid Res. 62 (2021), 100033 . \n",
      "[161] E. Cianci, A. Recchiuti, O. Trubiani, F. Diomede, M. Marchisio, S. Miscia, R. \n",
      "A. Colas, J. Dalli, C.N. Serhan, M. Romano, Human periodontal stem cells release \n",
      "specialized proresolving mediators and carry immunomodulatory and prohealing \n",
      "properties regulated by lipoxins, Stem Cells Transl. Med. 5 (1) (2016) 20 – 32 . \n",
      "[162] T.L. Pham, A.H. Kakazu, J. He, R. Nshimiyimana, N.A. Petasis, B. Jun, N.G. Bazan, \n",
      "H.E.P. Bazan, Elucidating the structure and functions of Resolvin D6 isomers on \n",
      "nerve regeneration with a distinctive trigeminal transcriptome, FASEB J. 35 (8) \n",
      "(2021), e21775 . \n",
      "[163] S. Hong, E. Tjonahen, E.L. Morgan, L. Yu, C.N. Serhan, A.F. Rowley, Rainbow \n",
      "trout ( Oncorhynchus mykiss ) brain cells biosynthesize novel docosahexaenoic acid- \n",
      "derived resolvins and protectins – mediator lipidomic analysis, Prostaglandins \n",
      "Other Lipid Mediat. 78 (2005) 107 – 116 . \n",
      "[164] P.C. Calder, Eicosapentaenoic and docosahexaenoic acid derived specialised pro- \n",
      "resolving mediators: concentrations in humans and the effects of age, sex, disease \n",
      "and increased omega-3 fatty acid intake, Biochimie 178 (2020) 105 – 123 . \n",
      "[165] S.K. Raatz, M.Y. Golovko, S.A. Brose, T.A. Rosenberger, G.S. Burr, W.R. Wolters, \n",
      "M.J. Picklo Sr., Baking reduces prostaglandin, resolvin, and hydroxy-fatty acid \n",
      "content of farm-raised Atlantic salmon (Salmo salar), J. Agric. Food Chem. 59 \n",
      "(2011) 11278 – 11286 . \n",
      "[166] P.K. Mukherjee, V.L. Marcheselli, C.N. Serhan, N.G. Bazan, Neuroprotectin D1: a \n",
      "docosahexaenoic acid-derived docosatriene protects human retinal pigment \n",
      "epithelial cells from oxidative stress, Proc. Natl. Acad. Sci. U. S. A. 101 (2004) \n",
      "8491 – 8496 . \n",
      "[167] V.L. Marcheselli, S. Hong, W.J. Lukiw, X. Hua Tian, K. Gronert, A. Musto, \n",
      "M. Hardy, J.M. Gimenez, N. Chiang, C.N. Serhan, N.G. Bazan, Novel docosanoids \n",
      "inhibit brain ischemia-reperfusion-mediated leukocyte infiltration and pro- \n",
      "inflammatory gene expression, J. Biol. Chem. 278 (2003) 43807 – 43817 . \n",
      "[168] K. Gronert, N. Maheshwari, N. Khan, I.R. Hassan, M. Dunn, M.L. Schwartzman, \n",
      "A role for the mouse 12/15-lipoxygenase pathway in promoting epithelial wound \n",
      "healing and host defense, J. Biol. Chem. 280 (2005) 15267 – 15278 . \n",
      "[169] B.D. Levy, P. Kohli, K. Gotlinger, O. Haworth, S. Hong, S. Kazani, E. Israel, K. \n",
      "J. Haley, C.N. Serhan, Protectin D1 is generated in asthma and dampens airway \n",
      "inflammation and hyper-responsiveness, J. Immunol. 178 (2007) 496 – 502 . \n",
      "[170] J.S. Duffield, S. Hong, V. Vaidya, Y. Lu, G. Fredman, C.N. Serhan, J.V. Bonventre, \n",
      "Resolvin D series and protectin D1 mitigate acute kidney injury, J. Immunol. 177 \n",
      "(2006) 5902 – 5911 . \n",
      "[171] W.J. Lukiw, J.G. Cui, V.L. Marcheselli, M. Bodker, A. Botkjaer, K. Gotlinger, C. \n",
      "N. Serhan, N.G. Bazan, A role for docosahexaenoic acid-derived neuroprotectin \n",
      "D1 in neural cell survival and Alzheimer disease, J. Clin. Invest. 115 (2005) \n",
      "2774 – 2783 . \n",
      "[172] C.N. Serhan, K. Gotlinger, S. Hong, Y. Lu, J. Siegelman, T. Baer, R. Yang, S. \n",
      "P. Colgan, N.A. Petasis, Anti-inflammatory actions of neuroprotectin D1/protectin \n",
      "D1 and its natural stereoisomers: assignments of dihydroxy-containing \n",
      "docosatrienes, J. Immunol. 176 (2006) 1848 – 1859 . \n",
      "[173] C.N. Serhan, G. Fredman, R. Yang, S. Karamnov, L.S. Belayev, N.G. Bazan, M. Zhu, \n",
      "J.W. Winkler, N.A. Petasis, Novel proresolving aspirin-triggered DHA pathway, \n",
      "Chem. Biol. 18 (2011) 976 – 987 . \n",
      "[174] C.N. Serhan, J. Dalli, R.A. Colas, J.W. Winkler, N. Chiang, Protectins and \n",
      "maresins: new pro-resolving families of mediators in acute inflammation and \n",
      "resolution bioactive metabolome, Biochim. Biophys. Acta 1851 (2015) 397 – 413 . \n",
      "[175] C.N. Serhan, J. Dalli, S. Karamnov, A. Choi, C.K. Park, Z.Z. Xu, R.R. Ji, M. Zhu, N. \n",
      "A. Petasis, Macrophage pro-resolving mediator maresin 1 stimulates tissue \n",
      "regeneration and controls pain, FASEB J. 26 (2012) 1755 – 1765 . \n",
      "[176] J. Dalli, M. Zhu, N.A. Vlasenko, B. Deng, J.Z. Haeggstrom, N.A. Petasis, C. \n",
      "N. Serhan, The novel 13S,14S-epoxy-maresin is converted by human \n",
      "macrophages to maresin1 (MaR1), inhibits leukotriene A4 hydrolase (LTA4H), \n",
      "and shifts macrophage phenotype, FASEB J. 27 (2013) 2573 – 2583 . \n",
      "[177] B. Deng, C.W. Wang, H.H. Arnardottir, Y. Li, C.Y. Cheng, J. Dalli, C.N. Serhan, \n",
      "Maresin biosynthesis and identification of maresin 2, a new anti-inflammatory \n",
      "and pro-resolving mediator from human macrophages, PLoS One 9 (7) (2014), \n",
      "e102362 . \n",
      "[178] S. Ramon, J. Dalli, J.M. Sanger, J.W. Winkler, M. Aursnes, J.E. Tungen, T. \n",
      "V. Hansen, C.N. Serhan, The Protectin PCTR1 is produced by human M2 \n",
      "macrophages and enhances resolution of infectious inflammation, Am. J. Pathol. \n",
      "186 (2016) 962 – 973 . \n",
      "[179] J. Dalli, N. Chiang, C.N. Serhan, Identification of sulfido-conjugated mediators \n",
      "that promote resolution of infection and organ protection, Proc. Natl. Acad. Sci. \n",
      "U. S. A. 111 (2014) E4753 – 4761 . \n",
      "[180] X. de la Rosa, P.C. Norris, N. Chiang, A.R. Rodriguez, B.W. Spur, C.N. Serhan, \n",
      "Identification and complete stereochemical assignments of the new Resolvin \n",
      "Conjugates in Tissue Regeneration (RCTR) in human tissues that stimulate \n",
      "proresolving phagocyte functions and tissue regeneration, Am. J. Pathol. 188 \n",
      "(2018) 950 – 966 . \n",
      "[181] J. Dalli, S. Ramon, P.C. Norris, R.A. Colas, C.N. Serhan, Novel proresolving and \n",
      "tissue regenerative resolvin and protectin sulfido-conjugated pathways, FASEB J. \n",
      "29 (2015) 2120 – 2136 . \n",
      "[182] J. Dalli, I. Vlasakov, I.R. Riley, A.R. Rodriguez, B. Spur, N.A. Petasis, N. Chiang, C. \n",
      "N. Serhan, Maresin conjugates in tissue regeneration biosynthesis enzymes in \n",
      "human macrophages, Proc. Natl. Acad. Sci. U. S. A. 113 (2016) 12232 – 12237 . \n",
      "[183] C.C. Jouvene, A.E. Shay, M.A. Soens, P.C. Norris, J.Z. Haeggstrom, C.N. Serhan, \n",
      "Biosynthetic metabolomes of cysteinyl-containing immunoresolvents, FASEB J. \n",
      "33 (2019) 13794 – 13807 . \n",
      "[184] N. Chiang, C.N. Serhan, Specialized pro-resolving mediator network: an update \n",
      "on production and actions, Essays Biochem. 64 (2020) 443 – 462 . \n",
      "[185] K.R. Maddipati, Non-inflammatory physiology of “ Inflammatory ” mediators - \n",
      "unalamation, a new paradigm, Front. Immunol. 11 (2020), 580117 . \n",
      "[186] J. So, D. Wu, A.H. Lichtenstein, A.K. Tai, N.R. Matthan, K.R. Maddipati, S. Lamon- \n",
      "Fava, EPA and DHA differentially modulate monocyte inflammatory response in \n",
      "subjects with chronic inflammation in part via plasma specialized pro-resolving \n",
      "lipid mediators: a randomized, double-blind, crossover study, Atherosclerosis 316 \n",
      "(2021) 90 – 98 . \n",
      "[187] W.C. Tsai, C. Kalyanaraman, A. Yamaguchi, M. Holinstat, M.P. Jacobson, T. \n",
      "R. Holman, In vitro biosynthetic pathway investigations of neuroprotectin D1 \n",
      "(NPD1) and protectin DX (PDX) by human 12-lipoxygenase, 15-lipoxygenase-1, \n",
      "and 15-lipoxygenase-2, Biochemistry 60 (22) (2021) 1741 – 1754 . \n",
      "C.N. Serhan et al.                                                                                                                                                                                                                               \n",
      "Seminars in Immunology 59 (2022) 101597\n",
      "17\n",
      "[188] C.R. Lee, D.C. Zeldin, Resolvin infectious inflammation by targeting the host \n",
      "response, N. Engl. J. Med. 373 (22) (2015) 2183 – 2185 . \n",
      "[189] C.N. Serhan, N. Chiang, J. Dalli, B.D. Levy, Lipid mediators in the resolution of \n",
      "inflammation, Cold Spring Harb. Perspect. Biol. 7 (2) (2014), a016311 . \n",
      "[190] A.P. Simopoulos, C.N. Serhan, R.P. Bazinet, The need for precision nutrition, \n",
      "genetic variation and resolution in Covid-19 patients, Mol. Aspects Med. 77 \n",
      "(2021), 100943 . \n",
      "[191] L.V. Norling, M. Spite, R. Yang, R.J. Flower, M. Perretti, C.N. Serhan, Cutting \n",
      "edge: humanized nano-proresolving medicines mimic inflammation-resolution \n",
      "and enhance wound healing, J. Immunol. 186 (2011) 5543 – 5547 . \n",
      "[192] M.C.P. Sok, M.C. Tria, C.E. Olingy, C.L. San Emeterio, E.A. Botchwey, Aspirin- \n",
      "Triggered Resolvin D1-modified materials promote the accumulation of pro- \n",
      "regenerative immune cell subsets and enhance vascular remodeling, Acta \n",
      "Biomater. 53 (2017) 109 – 122 . \n",
      "[193] R. Menon, P. Krzyszczyk, F. Berthiaume, Pro-resolution potency of resolvins D1, \n",
      "D2 and E1 on neutrophil migration and in dermal wound healing, Nano Life 7 (1) \n",
      "(2017), 1750002 . \n",
      "[194] P.C. Norris, S. Libreros, N. Chiang, C.N. Serhan, A cluster of immunoresolvents \n",
      "links coagulation to innate host defense in human blood, Sci. Signal. 10 (2017), \n",
      "eaan1471 . \n",
      "[195] H. Arnardottir, S.C. Pawelzik, U. \n",
      "¨\n",
      "Ohlund Wistbacka, G. Artiach, R. Hofmann, \n",
      "I. Reinholdsson, F. Braunschweig, P. Tornvall, D. Religa, M. B ¨ack, Stimulating the \n",
      "resolution of inflammation through Omega-3 polyunsaturated fatty acids in \n",
      "COVID-19: rationale for the COVID-Omega-F trial, Front. Physiol. 11 (2020), \n",
      "624657 . \n",
      "[196] D.W. Gilroy, T. Lawrence, M. Perretti, A.G. Rossi, Inflammatory resolution: new \n",
      "opportunities for drug discovery, Nat. Rev. Drug Discov. 3 (2004) 401 – 416 . \n",
      "[197] C.N. Serhan, J. Savill, Resolution of inflammation: the beginning programs the \n",
      "end, Nat. Immunol. 6 (2005) 1191 – 1197 . \n",
      "[198] G. Fredman, T.E. Van Dyke, C.N. Serhan, Resolvin E1 regulates adenosine \n",
      "diphosphate activation of human platelets, Arterioscler. Thromb. Vasc. Biol. 30 \n",
      "(10) (2010) 2005 – 2013 . \n",
      "[199] B.D. Levy, Q.Y. Zhang, C. Bonnans, V. Primo, J.J. Reilly, D.L. Perkins, Y. Liang, \n",
      "M. Amin Arnaout, B. Nikolic, C.N. Serhan, The endogenous pro-resolving \n",
      "mediators lipoxin A4 and resolvin E1 preserve organ function in allograft \n",
      "rejection, Prostaglandins Leukot. Essent. Fatty Acids 84 (1 – 2) (2011) 43 – 50 . \n",
      "[200] A. Kantarci, N. Aytan, I. Palaska, D. Stephens, L. Crabtree, C. Benincasa, B. \n",
      "G. Jenkins, I. Carreras, A. Dedeoglu, Combined administration of resolvin E1 and \n",
      "lipoxin A4 resolves inflammation in a murine model of Alzheimer ’ s disease, Exp. \n",
      "Neurol. 300 (2018) 111 – 120 . \n",
      "[201] A. Laguna-Fernandez, A. Checa, M. Carracedo, G. Artiach, M.H. Petri, \n",
      "R. Baumgartner, M.J. Forteza, X. Jiang, T. Andonova, M.E. Walker, J. Dalli, \n",
      "H. Arnardottir, A. Gistera, S. Thul, C.E. Wheelock, G. Paulsson-Berne, D.F. \n",
      "J. Ketelhuth, G.K. Hansson, M. Back, ERV1/ChemR23 signaling protects against \n",
      "atherosclerosis by modifying oxidized low-density lipoprotein uptake and \n",
      "phagocytosis in macrophages, Circulation 138 (16) (2018) 1693 – 1705 . \n",
      "[202] S.P. Colgan, Resolvins resolve to heal mucosal wounds, Proc. Natl. Acad. Sci. U. S. \n",
      "A. 117 (20) (2020) 10621 – 10622 . \n",
      "[203] Y. Isobe, M. Arita, R. Iwamoto, D. Urabe, H. Todoroki, K. Masuda, M. Inoue, \n",
      "H. Arai, Stereochemical assignment and anti-inflammatory properties of the \n",
      "omega-3 lipid mediator resolvin E3, J. Biochem. 153 (4) (2013) 355 – 360 . \n",
      "[204] J. Li, C.Y. Chen, M. Arita, K. Kim, X. Li, H. Zhang, J.X. Kang, An omega-3 \n",
      "polyunsaturated fatty acid derivative, 18-HEPE, protects against CXCR4- \n",
      "associated melanoma metastasis, Carcinogenesis 39 (11) (2018) 1380 – 1388 . \n",
      "[205] S. Fiore, C.N. Serhan, Formation of lipoxins and leukotrienes during receptor- \n",
      "mediated interactions of human platelets and recombinant human granulocyte/ \n",
      "macrophage colony-stimulating factor-primed neutrophils, J. Exp. Med. 172 (5) \n",
      "(1990) 1451 – 1457 . \n",
      "[206] A.F. Rowley, P. Lloyd-Evans, S.E. Barrow, C.N. Serhan, Lipoxin biosynthesis by \n",
      "trout macrophages involves the formation of epoxide intermediates, Biochemistry \n",
      "33 (4) (1994) 856 – 863 . \n",
      "[207] O. Koltsida, S. Karamnov, K. Pyrillou, T. Vickery, A.D. Chairakaki, \n",
      "C. Tamvakopoulos, P. Sideras, C.N. Serhan, E. Andreakos, Toll-like receptor 7 \n",
      "stimulates production of specialized pro-resolving lipid mediators and promotes \n",
      "resolution of airway inflammation, EMBO Mol. Med. 5 (5) (2013) 762 – 775 . \n",
      "[208] A. Lukic, P. Larssen, A. Fauland, B. Samuelsson, C.E. Wheelock, S. Gabrielsson, \n",
      "O. Radmark, GM-CSF- and M-CSF-primed macrophages present similar resolving \n",
      "but distinct inflammatory lipid mediator signatures, FASEB J. 31 (10) (2017) \n",
      "4370 – 4381 . \n",
      "[209] T. Yamada, Y. Tani, H. Nakanishi, R. Taguchi, M. Arita, H. Arai, Eosinophils \n",
      "promote resolution of acute periotonitis by producing proresolving mediators in \n",
      "mice, FASEB J. 25 (2011) 561 – 568 . \n",
      "[210] N. Chiang, M. Shinohara, J. Dalli, V. Mirakaj, M. Kibi, A.M. Choi, C.N. Serhan, \n",
      "Inhaled carbon monoxide accelerates resolution of inflammation via unique \n",
      "proresolving mediator-heme oxygenase-1 circuits, J. Immunol. 190 (12) (2013) \n",
      "6378 – 6388 . \n",
      "[211] S.-E. Dahl ´en, C.N. Serhan, Lipoxins: bioactive lipoxygenase interaction products, \n",
      "in: A. Wong, S.T. Crooke (Eds.), Lipoxygenases and Their Products, Academic \n",
      "Press, San Diego, 1991, pp. 235 – 276 . \n",
      "[212] M. Ogawa, T. Ishihara, Y. Isobe, T. Kato, K. Kuba, Y. Imai, Y. Uchino, K. Tsubota, \n",
      "M. Arita, Eosinophils promote corneal wound healing via the 12/15-lipoxygenase \n",
      "pathway, FASEB J. 34 (9) (2020) 12492 – 12501 . \n",
      "[213] M. Arita, T. Ohira, Y.P. Sun, S. Elangovan, N. Chiang, C.N. Serhan, Resolvin E1 \n",
      "selectively interacts with leukotriene B\n",
      "4 \n",
      "receptor BLT1 and ChemR23 to regulate \n",
      "inflammation, J. Immunol. 178 (2007) 3912 – 3917 . \n",
      "[214] S. Hong, T.F. Porter, Y. Lu, S.F. Oh, P.S. Pillai, C.N. Serhan, Resolvin E1 \n",
      "metabolome in local inactivation during inflammation-resolution, J. Immunol. \n",
      "180 (2008) 3512 – 3519 . \n",
      "C.N. Serhan et al.                                                                                                                                                                                                                               \n"
     ]
    }
   ],
   "source": [
    "for l in LC_69_pages[13:]:\n",
    "    print(l.page_content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0af9cf5c-869d-4ba4-b734-683218f9d348",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of chunks: 62\n",
      "Processing chunk #0\n",
      "Processing chunk #1\n",
      "Processing chunk #2\n",
      "Processing chunk #3\n",
      "Processing chunk #4\n",
      "Processing chunk #5\n",
      "Processing chunk #6\n",
      "Processing chunk #7\n",
      "Processing chunk #8\n",
      "Processing chunk #9\n",
      "Processing chunk #10\n",
      "Processing chunk #11\n",
      "Processing chunk #12\n",
      "Processing chunk #13\n",
      "Processing chunk #14\n",
      "Processing chunk #15\n",
      "Processing chunk #16\n",
      "Processing chunk #17\n",
      "Processing chunk #18\n",
      "Processing chunk #19\n",
      "Processing chunk #20\n",
      "Processing chunk #21\n",
      "Processing chunk #22\n",
      "Processing chunk #23\n",
      "Processing chunk #24\n",
      "Processing chunk #25\n",
      "Processing chunk #26\n",
      "Processing chunk #27\n",
      "Processing chunk #28\n",
      "Processing chunk #29\n",
      "Processing chunk #30\n",
      "Processing chunk #31\n",
      "Processing chunk #32\n",
      "Processing chunk #33\n",
      "Processing chunk #34\n",
      "Processing chunk #35\n",
      "Processing chunk #36\n",
      "Processing chunk #37\n",
      "Processing chunk #38\n",
      "Processing chunk #39\n",
      "Processing chunk #40\n",
      "Processing chunk #41\n",
      "Processing chunk #42\n",
      "Processing chunk #43\n",
      "Processing chunk #44\n",
      "Processing chunk #45\n",
      "Processing chunk #46\n",
      "Processing chunk #47\n",
      "Processing chunk #48\n",
      "Processing chunk #49\n",
      "Processing chunk #50\n",
      "Processing chunk #51\n",
      "Processing chunk #52\n",
      "Processing chunk #53\n",
      "Processing chunk #54\n",
      "Processing chunk #55\n",
      "Processing chunk #56\n",
      "Processing chunk #57\n",
      "Processing chunk #58\n",
      "Processing chunk #59\n",
      "Processing chunk #60\n",
      "Processing chunk #61\n",
      "Number of chunks: 62\n",
      "Processing chunk #0\n",
      "Processing chunk #1\n",
      "Processing chunk #2\n",
      "Processing chunk #3\n",
      "Processing chunk #4\n",
      "Processing chunk #5\n",
      "Processing chunk #6\n",
      "Processing chunk #7\n",
      "Processing chunk #8\n",
      "Processing chunk #9\n",
      "Processing chunk #10\n",
      "Processing chunk #11\n",
      "Processing chunk #12\n",
      "Processing chunk #13\n",
      "Processing chunk #14\n",
      "Processing chunk #15\n",
      "Processing chunk #16\n",
      "Processing chunk #17\n",
      "Processing chunk #18\n",
      "Processing chunk #19\n",
      "Processing chunk #20\n",
      "Processing chunk #21\n",
      "Processing chunk #22\n",
      "Processing chunk #23\n",
      "Processing chunk #24\n",
      "Processing chunk #25\n",
      "Processing chunk #26\n",
      "Processing chunk #27\n",
      "Processing chunk #28\n",
      "Processing chunk #29\n",
      "Processing chunk #30\n",
      "Processing chunk #31\n",
      "Processing chunk #32\n",
      "Processing chunk #33\n",
      "Processing chunk #34\n",
      "Processing chunk #35\n",
      "Processing chunk #36\n",
      "Processing chunk #37\n",
      "Processing chunk #38\n",
      "Processing chunk #39\n",
      "Processing chunk #40\n",
      "Processing chunk #41\n",
      "Processing chunk #42\n",
      "Processing chunk #43\n",
      "Processing chunk #44\n",
      "Processing chunk #45\n",
      "Processing chunk #46\n",
      "Processing chunk #47\n",
      "Processing chunk #48\n",
      "Processing chunk #49\n",
      "Processing chunk #50\n",
      "Processing chunk #51\n",
      "Processing chunk #52\n",
      "Processing chunk #53\n",
      "Processing chunk #54\n",
      "Processing chunk #55\n",
      "Processing chunk #56\n",
      "Processing chunk #57\n",
      "Processing chunk #58\n",
      "Processing chunk #59\n",
      "Processing chunk #60\n",
      "Processing chunk #61\n",
      "Number of chunks: 31\n",
      "Processing chunk #0\n",
      "Processing chunk #1\n",
      "Processing chunk #2\n",
      "Processing chunk #3\n",
      "Processing chunk #4\n",
      "Processing chunk #5\n",
      "Processing chunk #6\n",
      "Processing chunk #7\n",
      "Processing chunk #8\n",
      "Processing chunk #9\n",
      "Processing chunk #10\n",
      "Processing chunk #11\n",
      "Processing chunk #12\n",
      "Processing chunk #13\n",
      "Processing chunk #14\n",
      "Processing chunk #15\n",
      "Processing chunk #16\n",
      "Processing chunk #17\n",
      "Processing chunk #18\n",
      "Processing chunk #19\n",
      "Processing chunk #20\n",
      "Processing chunk #21\n",
      "Processing chunk #22\n",
      "Processing chunk #23\n",
      "Processing chunk #24\n",
      "Processing chunk #25\n",
      "Processing chunk #26\n",
      "Processing chunk #27\n",
      "Processing chunk #28\n",
      "Processing chunk #29\n",
      "Processing chunk #30\n",
      "CPU times: user 3.58 s, sys: 294 ms, total: 3.87 s\n",
      "Wall time: 22min 27s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "keywords_all_pages_no_omit = extract_article_keywords(LC_69_pages, model=LLM, omit_bibliography=False)\n",
    "keywords_all_pages_omit = extract_article_keywords(LC_69_pages, model=LLM, omit_bibliography=True)\n",
    "keywords_up_to_biblio = extract_article_keywords(pages_up_to_bibliography, model=LLM, omit_bibliography=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "07b15be9-b893-442d-889f-fac96c9d2114",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('COVID-19', 'COVID-19'),\n",
       " ('COVID-19', 'COVID-19'),\n",
       " ('COVID-19', 'COVID-19'),\n",
       " ('COVID-19', 'COVID-19'),\n",
       " ('acute inflammation', 'COVID-19'),\n",
       " ('collateral organ damage', 'COVID-19'),\n",
       " ('tissue injury', 'gingival inflammation'),\n",
       " ('COVID-19', 'inflammatory response'),\n",
       " ('COVID-19', 'COVID-19'),\n",
       " ('gingival inflammation', 'COVID-19'),\n",
       " ('cardiovascular disease', 'hemorrhagic exudate'),\n",
       " ('COVID-19', 'COVID-19'),\n",
       " ('human rhinosinusitis', 'COVID-19'),\n",
       " ('kidney diseases', 'unknown'),\n",
       " ('diabetic complications', 'peritonitis'),\n",
       " ('COVID-19', 'sepsis'),\n",
       " ('COVID-19', 'ischemia-reperfusion injury'),\n",
       " ('Cancer', 'diabetes'),\n",
       " ('hemorrhagic exudate', 'colitis'),\n",
       " ('COVID-19', 'lung inflammation'),\n",
       " ('COVID-19', 'obesity'),\n",
       " ('unknown', 'atherosclerosis'),\n",
       " ('peritonitis', 'tumor burden'),\n",
       " ('sepsis', 'dermatitis'),\n",
       " ('ischemia-reperfusion injury', 'Candida albicans infection'),\n",
       " ('diabetes', 'herpes simplex virus infection'),\n",
       " ('colitis', 'metabolic syndrome'),\n",
       " ('lung inflammation', 'arthritis'),\n",
       " ('obesity', 'peripheral artery disease'),\n",
       " ('atherosclerosis', 'type II diabetes mellitus'),\n",
       " ('tumor burden', 'peripheral artery disease'),\n",
       " ('dermatitis', 'COVID-19'),\n",
       " ('Candida albicans infection', 'depression'),\n",
       " ('herpes simplex virus infection', 'cardiovascular disease'),\n",
       " ('metabolic syndrome', 'peripheral artery disease'),\n",
       " ('type II diabetes mellitus', 'coronary artery disease'),\n",
       " ('peripheral artery disease', 'coronary microvascular dysfunction'),\n",
       " ('COVID-19', 'peritonitis'),\n",
       " ('cardiovascular disease', 'diabetes'),\n",
       " ('peripheral artery disease', 'obesity'),\n",
       " ('peritonitis', 'colitis'),\n",
       " ('diabetes', 'lung inflammation'),\n",
       " ('obesity', 'kidney injury'),\n",
       " ('colitis', 'depression'),\n",
       " ('lung inflammation', \"Alzheimer's\"),\n",
       " ('kidney injury', 'atherosclerosis'),\n",
       " ('depression', 'dermatitis'),\n",
       " (\"Alzheimer's\", 'pain'),\n",
       " ('atherosclerosis', 'herpes'),\n",
       " ('dermatitis', 'periodontal ligaments'),\n",
       " ('pain', 'metabolic syndrome'),\n",
       " ('periodontal ligaments', 'Peritonitis'),\n",
       " ('metabolic syndrome', 'Arthritis'),\n",
       " ('arthritis', 'Type 2 diabetes mellitus'),\n",
       " ('peripheral artery disease', 'Peripheral artery disease'),\n",
       " ('Peritonitis', 'Premature birth'),\n",
       " ('Arthritis', 'Depression'),\n",
       " ('Type 2 diabetes mellitus', 'Diabetes'),\n",
       " ('Peripheral artery disease', 'diabetes'),\n",
       " ('Murine depression behavior', 'COVID-19'),\n",
       " ('Premature birth', 'abdominal septic shock'),\n",
       " ('diabetes', 'experimental arthritis'),\n",
       " ('diabetes', 'muscle inflammation'),\n",
       " ('psoriasis', 'nerve regeneration'),\n",
       " ('COVID-19', 'wound healing'),\n",
       " ('neuropathic pain', 'neuropathic pain'),\n",
       " ('abdominal septic shock', 'metabolic syndrome'),\n",
       " ('muscle inflammation', 'neurodegenerative diseases'),\n",
       " ('tendon overuse', 'unknown'),\n",
       " ('metabolic syndrome', 'Infectious Leukocytes'),\n",
       " ('neurodegenerative diseases', 'pro-inflammatory conditions'),\n",
       " ('unknown', 'COVID-19'),\n",
       " ('unknown', 'gingival inflammation'),\n",
       " ('COVID-19', None),\n",
       " ('COVID-19', None),\n",
       " ('long COVID', None),\n",
       " ('acute inflammation', None),\n",
       " ('COVID-19', None),\n",
       " ('COVID-19', None),\n",
       " ('COVID-19', None),\n",
       " ('COVID-19', None),\n",
       " ('chronic kidney disease', None),\n",
       " ('obesity', None),\n",
       " ('cystic fibrosis', None),\n",
       " ('Type 2 diabetes mellitus', None),\n",
       " ('major depressive disorder', None),\n",
       " ('obesity', None),\n",
       " ('diabetes', None),\n",
       " ('diabetic kidney disease', None),\n",
       " ('chronic rhinosinusitis', None),\n",
       " ('COVID-19-associated cardiovascular complications', None),\n",
       " ('chronic inflammation', None),\n",
       " ('periodontitis', None),\n",
       " (\"Alzheimer's disease\", None),\n",
       " (\"Alzheimer's disease\", None),\n",
       " ('Colitis', None),\n",
       " ('Periodontitis', None),\n",
       " ('polymicrobial sepsis', None),\n",
       " ('type 2 diabetes', None),\n",
       " ('obesity-induced insulin resistance', None),\n",
       " ('hepatic steatosis', None),\n",
       " ('peripheral atherosclerosis', None),\n",
       " ('psoriatic dermatitis', None),\n",
       " ('ocular inflammatory lesions', None),\n",
       " ('systemic inflammation', None),\n",
       " ('cardiovascular disease', None),\n",
       " ('cardiovascular disease', None),\n",
       " ('allergic airway inflammation', None),\n",
       " ('coronary microvascular dysfunction', None),\n",
       " ('psoriasis', None),\n",
       " ('experimental arthritis', None),\n",
       " ('Asthma', None),\n",
       " ('Alzheimer disease', None),\n",
       " ('Acute kidney injury', None),\n",
       " ('chronic inflammation', None),\n",
       " ('COVID-19', None),\n",
       " (\"Alzheimer's disease\", None),\n",
       " ('melanoma metastasis', None),\n",
       " ('unknown', None)]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import itertools\n",
    "list(\n",
    "    itertools.zip_longest(\n",
    "        map(lambda d: d.name, keywords_all_pages_omit.mentioned_diseases),\n",
    "        map(lambda d: d.name, keywords_up_to_biblio.mentioned_diseases),\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0efbcadd-3800-4915-a3d2-2fd32ae12e27",
   "metadata": {},
   "source": [
    "# Compare ground truth with MolGenetMetab results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "e8a05642-1aa9-4123-b67b-2c10f87cfdf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "ground_truth_metabolites = \"\"\"formate\n",
    "sarcosine\n",
    "hypoxanthine\n",
    "homocysteine\n",
    "ATP\n",
    "NAD\n",
    "NADH\n",
    "fatty acids\n",
    "dihydroceramide\n",
    "unsaturated fatty acid\n",
    "phosphocholines\n",
    "cholesterol ester\n",
    "hydroxypropionylcarnitine\n",
    "acetate\n",
    "dodecanedioic acid\n",
    "indoxylsulfate\n",
    "arginine\n",
    "glucose\n",
    "carnitine\n",
    "glutamate\n",
    "lysine\n",
    "histidine\n",
    "branched chain amino acids\n",
    "leucine\n",
    "valine\n",
    "choline\n",
    "threonine\n",
    "ornithine\n",
    "lactate\n",
    "succinate\n",
    "cholesterol ester 20:0\n",
    "triglycerides\n",
    "phosphatidylcholine\n",
    "betaine\n",
    "methionine\n",
    "s-adenosylmethionine\n",
    "s-adenosylhomocysteine\n",
    "dimethylglycine\n",
    "iron-sulfur clusters\n",
    "phosphocreatine\n",
    "nicotinamide mononucleotide\n",
    "oxylipin\n",
    "reactive oxygen species\n",
    "taurine\n",
    "β-alanine\n",
    "pyruvate\n",
    "\"\"\".strip().split(\"\\n\")\n",
    "\n",
    "ground_truth_pathways = \"\"\"iron-sulfur cluster biogenesis\n",
    "cellular energy metabolism\n",
    "mitochondrial electron transport chain\n",
    "krebs cycle\n",
    "electron transfer flavoprotein\n",
    "nuclear gene expression\n",
    "glycolysis\n",
    "carbohydrate and fatty acid metabolism\n",
    "energy metabolism\n",
    "one-carbon metabolism\n",
    "folate cycle\n",
    "methionine salvage\n",
    "purine nucleotide salvage and synthesis\n",
    "pyruvate metabolism\n",
    "\"\"\".strip().split(\"\\n\")\n",
    "\n",
    "ground_truth_proteins = [\n",
    "    \"frataxin\",\n",
    "]\n",
    "\n",
    "ground_truth_drugs = [\n",
    "    \"Etravirine\",\n",
    "    \"Resveratrol\",\n",
    "    \"SS-31\",\n",
    "    \"deferoxamine\",\n",
    "    \"BAPTA-AM\",\n",
    "    \"antioxidants\",\n",
    "]\n",
    "\n",
    "ground_truth_diseases = [\n",
    "    \"friedreich ataxia\",\n",
    "    \"dyslipidemia\",\n",
    "    \"pre-diabetic state\",\n",
    "    \"diabetes\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "98a0b284-b0b6-4bd4-89a6-c8d7050e8f87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# RUN 1\n",
      "\n",
      "## Metabolites:\n",
      "\n",
      "TP=38, FP=34, FN=8\n",
      "\n",
      "FP: tg(18:1,26:0), hydrophilic metabolites, 10-formyl-thf, ribose-5-phosphate, adenosylmethionine, s-adenosylmethionine (sam), kynurenine, serine, histamine, dodecanedioc acid, inosine monophosphate (imp), hypoxanthine (hxan), glycine, 5,10-methylene thf, prostaglandins, sam, edta, s-adenosylhomocysteine (sah), sah, oxylipins, creatine, lipid-related metabolites, 5,10-methylene-thf, betaine aldehyde, iron-sulfur (fe/s) cluster, uric acid, xanthine, tetrahydrofolate, hydroxypropionylcarnitine (c3.oh), inosine monophosphate, iron, tetrahydrofolate (thf), nicotinamide mononucleotide (nmn), uric acid (ua)\n",
      "\n",
      "FN: dodecanedioic acid, fatty acids, iron-sulfur clusters, phosphatidylcholine, unsaturated fatty acid, branched chain amino acids, reactive oxygen species, oxylipin\n",
      "\n",
      "\n",
      "## Pathways:\n",
      "\n",
      "TP=7, FP=13, FN=7\n",
      "\n",
      "FP: non-folate-dependent pathways, fatty acid metabolism, glucose utilizing pathways, 1c metabolism pathway, carbohydrate metabolism, electron transport chain, purine salvage pathway, 1c metabolic pathway, 1c metabolic pathways, methionine salvage pathway, oxidative phosphorylation, oxidative metabolism, purine nucleotide metabolism\n",
      "\n",
      "FN: glycolysis, carbohydrate and fatty acid metabolism, pyruvate metabolism, electron transfer flavoprotein, iron-sulfur cluster biogenesis, cellular energy metabolism, nuclear gene expression\n",
      "\n",
      "\n",
      "## Drugs:\n",
      "\n",
      "TP=5, FP=6, FN=1\n",
      "\n",
      "FP: antioxidant treatment, deferiprone, peptide ss-31, β-blocker, selegiline, idebenone\n",
      "\n",
      "FN: antioxidants\n",
      "\n",
      "\n",
      "## Diseases:\n",
      "\n",
      "TP=2, FP=11, FN=2\n",
      "\n",
      "FP: metabolic profiling, leigh syndrome, mitochondrial disease, mitochondrial dysfunction, friedreich's ataxia (frda), friedreich's ataxia (fa), friedreich's ataxia, fa, serine one-carbon catabolism, fa (friedreich's ataxia), mitochondrial encephalopathy lactic acidosis and stroke-like episodes (melas)\n",
      "\n",
      "FN: dyslipidemia, pre-diabetic state\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "# RUN 2\n",
      "\n",
      "\n",
      "## Metabolites:\n",
      "\n",
      "TP=36, FP=34, FN=10\n",
      "\n",
      "FP: tg(18:1,26:0), 10-formyl-thf, ribose-5-phosphate, hydroxypropionlycarnitine, resveratrol, unknown, serine, kynurenine, phosphocholine, dodecanedioc acid, inosine monophosphate (imp), hypoxanthine (hxan), glycine, 5,10-methylene thf, thf, pcs, hxan, sam, 5,10-me-thf, hcys, sah, 5,10-methylene-thf, imp, xan, uric acid, xanthine, tetrahydrofolate, hydroxypropionylcarnitine (c3.oh), dmg, rib-5p, inosine monophosphate, ua, histamine, 5-me-thf\n",
      "\n",
      "FN: atp, dodecanedioic acid, fatty acids, iron-sulfur clusters, phosphatidylcholine, unsaturated fatty acid, phosphocreatine, branched chain amino acids, reactive oxygen species, oxylipin\n",
      "\n",
      "\n",
      "## Pathways:\n",
      "\n",
      "TP=5, FP=13, FN=9\n",
      "\n",
      "FP: non-folate-dependent pathways, one-carbon (1c) metabolism, methionine salvage pathway, peroxisomal β-oxidation, oxidative phosphorylation, oxidative metabolism, purine salvage pathway, unknown, purine nucleotide metabolism, folate metabolism, 1c metabolic pathways, purine nucleotide synthesis, 1c metabolism\n",
      "\n",
      "FN: glycolysis, carbohydrate and fatty acid metabolism, pyruvate metabolism, electron transfer flavoprotein, mitochondrial electron transport chain, energy metabolism, iron-sulfur cluster biogenesis, cellular energy metabolism, nuclear gene expression\n",
      "\n",
      "\n",
      "## Drugs:\n",
      "\n",
      "TP=4, FP=7, FN=2\n",
      "\n",
      "FP: antioxidant treatment, deferiprone, long-acting β-blocker, unknown, selegiline, peptide ss-31, idebenone\n",
      "\n",
      "FN: antioxidants, ss-31\n",
      "\n",
      "\n",
      "## Diseases:\n",
      "\n",
      "TP=2, FP=10, FN=2\n",
      "\n",
      "FP: fanconi anemia, leigh syndrome, mitochondrial encephalopathy lactic acidosis and stroke-like episodes (melas), mitochondrial disease, hereditary ataxias, friedreich's ataxia, fa, spastic paraplegias, fatty acid (fa) patients, fa patients\n",
      "\n",
      "FN: dyslipidemia, pre-diabetic state\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "def confusion_matrix(observed, expected):\n",
    "    true_pos = [obs for obs in observed if obs in expected]\n",
    "    false_pos = [obs for obs in observed if obs not in expected]\n",
    "    false_neg = [exp for exp in expected if exp not in observed]\n",
    "    return true_pos, false_pos, false_neg\n",
    "\n",
    "def print_confusion_matrix(cm):\n",
    "    tp, fp, fn = cm\n",
    "    print(f'TP={len(tp)}, FP={len(fp)}, FN={len(fn)}')\n",
    "\n",
    "def cmp_with_ground_truth(file: str):\n",
    "    with open(file) as fp:\n",
    "        results = json.load(fp)\n",
    "    metabolite_confusion = confusion_matrix(\n",
    "        set(m['name'].lower() for m in results['keywords']['mentioned_metabolites']),\n",
    "        set(m.lower() for m in ground_truth_metabolites),\n",
    "    )\n",
    "    pathway_confusion = confusion_matrix(\n",
    "        set(p['name'].lower() for p in results['keywords']['mentioned_pathways']),\n",
    "        set(p.lower() for p in ground_truth_pathways),\n",
    "    )\n",
    "    drug_confusion = confusion_matrix(\n",
    "        set(d['name'].lower() for d in results['keywords']['mentioned_drugs']),\n",
    "        set(d.lower() for d in ground_truth_drugs),\n",
    "    )\n",
    "    disease_confusion = confusion_matrix(\n",
    "        set(d['name'].lower() for d in results['keywords']['mentioned_diseases']),\n",
    "        set(d.lower() for d in ground_truth_diseases),\n",
    "    )\n",
    "\n",
    "    for name, cm in (('Metabolites', metabolite_confusion), ('Pathways', pathway_confusion), ('Drugs', drug_confusion), ('Diseases', disease_confusion)):\n",
    "        tp, fp, fn = cm\n",
    "        print(f'## {name}:\\n')\n",
    "        print_confusion_matrix(cm)\n",
    "        print(f'''\n",
    "FP: {\", \".join(fp)}\n",
    "\n",
    "FN: {\", \".join(fn)}\n",
    "\n",
    "''')\n",
    "\n",
    "print('# RUN 1\\n')\n",
    "cmp_with_ground_truth('../experiments/molgenetmetab-run1.json')\n",
    "\n",
    "print('\\n\\n# RUN 2\\n\\n')\n",
    "cmp_with_ground_truth('../experiments/molgenetmetab-run2.json')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
